---
title: "ewastools"
author: "Qianhui"
date: "21/11/2018"
output: html_document
---

This document is mainly for part9 of the GEO data set analysis, that is the quality control (check fetal sex & genotypes) of data sets from different tissue types. 

I used methods from `ewastools` package to identify samples with fetal sex misslabelled (check fetal sex), and also check the homogeneity of genotypes (using SNP probes to check genotype) for each sample. 

Fetal sex is estimated based on intensities from X/Y chromosomes. 

The theory of genotype checking is one sample (or one person/individual) could only have 3 status of DNA methylation (MM or MU or UU^*^) for a perticular SNP. So the density plot of SNP methylation should be a plot with 3 peaks. If there are many SNP probes with methylation between 0%-50% or 50%-100% for one sample, this sample is likely to have more technical variance (or potential contamination from other individual or othe tissue types) than other samples that have 3 peaks as expected.

^*^ MM: both allels were methylated (~100% methylation), often refer to AA or BB;   
    MU: one allel methylated, one allel unmethylated (~50% methylation), often refer to A/B genotype;   
    UU: both allels were unmethylated (~0% methylation), often refer to AA or BB.


```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.align = "center", error = TRUE, echo = TRUE) # fig.width = 7, 
```


# 1. load data and packages
```{r package, warning=FALSE, message=FALSE}

library(tidyverse)
library(minfi)
library(magrittr)
library(tibble)
library(data.table)
library(limma)
library(ewastools)
library(FactoMineR)
library(doParallel)
library(ggplot2)
library(ggpubr)
library(RColorBrewer)
pal <- brewer.pal(8, "Dark2")

# GEOmeta, GEO_phenotypes
load(file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_phenotypes_add10.RData")

GEO_phenotypes_Filtered <- readRDS(file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_phenotypes410_Filtered.rds")

GEO_phenotypes_Filtered <- GEO_phenotypes_Filtered[!(GEO_phenotypes_Filtered$Sample_name%in%c("GSM1617002_7668610068_R01C01","GSM1616993_7668610068_R06C02")),]

```

```{r}

GEO_baseDir <- file.path("/home/qianhui/DNAme/Process_decidual/GEO_dataSets/IDAT_all")

# get the RGset for all 392 450k samples
GEO_RGset_450k_Tissue <- read.metharray.exp(base = GEO_baseDir, 
                                        targets = GEO_phenotypes_Filtered[GEO_phenotypes_Filtered$ArrayType%in%"450K",], 
                                        extended = TRUE, force = TRUE)

# get the RGset for all 16 EPIC samples
GEO_RGset_EPIC_Tissue <- read.metharray.exp(base = GEO_baseDir, 
                                        targets = GEO_phenotypes_Filtered[GEO_phenotypes_Filtered$ArrayType%in%"EPIC",], 
                                        extended = TRUE, force = TRUE)

```


# 3. ewastools: check sex and contanmincation of sampels.

## Functions to use ewastools to check fetal sex and genotypes.

For snp_outliers, a metric assessing the outlierness of the SNP beta-values. High values may indicate either contaminated or failed samples.

### 1) `CheckFetalSex` Function
```{r checkFetalSex}
# let argument RGsetChara to be a charactor

CheckFetalSex <- function(RGsetChara){
  registerDoParallel(cores=10)
  # get the object from character
  RGset <- get(RGsetChara)
  # get meta data
  phenoData <- pData(RGset)
  
  # read idat files
  sampleDirs <- phenoData$filenames
  meth = read_idats(idat_files = sampleDirs,quiet=FALSE)

  # sex check
  ## dye bias corrected before sex check
  meth %<>% correct_dye_bias

  # phenoData to phenoData.dt
  phenoData.dt <- phenoData %>% as.data.frame() %>% rownames_to_column() %>% setDT()
  # add average total beta values for probes on X and Y chr
  phenoData.dt[ ,c("X","Y") := check_sex(meth)]
  
  # predict fetal sex
  phenoData.dt$predicted_sex = predict_sex(phenoData.dt$X,phenoData.dt$Y
                           # which(phenoData.dt$Fetal_Sex%in%'Male'),
                           # which(phenoData.dt$Fetal_Sex%in%'Female')
                           )

  phenoData.dt_v1 <- phenoData.dt %>%
    mutate(predicted_sex = str_replace_all(predicted_sex, c("m"="Male","f" ="Female")))
  
  phenoData.dt_v2 <- setDT(phenoData.dt_v1)
  
  # plot the X/Y chr intensities
  # it is better to put pch into 3 groups (red color marked the misslabelled samples)
  plot(Y ~ X,data=phenoData.dt_v2,
       pch=ifelse(phenoData.dt_v2$Fetal_Sex=="Female",1,
           ifelse(phenoData.dt_v2$Fetal_Sex=="Male", 4,13)),
       asp=1,xlab="Normalized X chromosome intensities",ylab="Normalized Y chromosome intensities")
  points(Y ~ X,data=phenoData.dt_v2[Fetal_Sex !=predicted_sex],
         pch=ifelse(phenoData.dt_v2[Fetal_Sex!=predicted_sex]$Fetal_Sex=="Female",1,
             ifelse(phenoData.dt_v2[Fetal_Sex!=predicted_sex]$Fetal_Sex=="Male",4,13)),
         col=2)
  legend("topright",pch=c(1,4,13),legend=c("Female","Male", "NA"))
  title(main = RGsetChara)
  
  # samples with fetal sex in meta data not same as predicted fetal sex
  phenoData.dt_v2[Fetal_Sex !=predicted_sex] %>% as_tibble %>% 
    dplyr::select(rowname:Study, X:predicted_sex) %>% print(n=Inf)
  
  # remove intermediat data
  rm(meth)
  
  # return phenoData_withPredict, the df with predicted fetal sex
  return(as_tibble(phenoData.dt_v2))
}

```


### Use `CheckFetalSex`
```{r CheckFetalSexforAllTissues}

TissueRGsetNames <- c("GEO_RGset_450k_Tissue", "GEO_RGset_EPIC_Tissue")


for(i in TissueRGsetNames){
  # check fetal sex for each subset data sets
  # and save the output df, the phenodata with predicted fetal sex
  assign(paste0("GEO_phenoData", str_replace(i, "GEO_RGset",""), "_PreFS"),   
         CheckFetalSex(RGsetChara = i))

}

# save all phenoData with predicted fetal sex infotmation

GEO_phenotypes_addEwastoolPredictedSex <- rbind(GEO_phenoData_450k_Tissue_PreFS,
                                        GEO_phenoData_EPIC_Tissue_PreFS)

saveRDS(GEO_phenotypes_addEwastoolPredictedSex,
  file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_phenotypes_addEwastoolPredictedSex.rds")

```


### Draw plot for sex and predicted sex

```{r}

# The decidua (EPIC array) sex info. is NA, which should be female, so we replace the sex info. from phenodata with female
GEO_phenotypes_addPredictedSex <- GEO_phenotypes_addEwastoolPredictedSex
GEO_phenotypes_addPredictedSex[is.na(GEO_phenotypes_addPredictedSex$Fetal_Sex), ]$Fetal_Sex <- rep("Female",2)

# add WrongSexInfo column
GEO_phenotypes_addPredictedSex$CorrectSexInfo <- GEO_phenotypes_addPredictedSex$Fetal_Sex == GEO_phenotypes_addPredictedSex$predicted_sex

# plot fetal sex
theme_set(theme_pubr())
GEO_phenotypes_addPredictedSex %>% 
  ggplot(aes(x=X, y=Y, colour=CorrectSexInfo)) +
  geom_point()
  

GEO_phenotypes_addPredictedSex %>% 
  dplyr::select(Sample, Sample_name, Trimester, Additional_Info, Fetal_Sex, Study, ArrayType, X, Y, predicted_sex, WrongSexInfo) %>% 
  dplyr::mutate(Outlier1=ifelse(.$Additional_Info%in%"Outlier", "Outlier", "Pure")) %>% 
  dplyr::mutate(Outlier = case_when(Sample_name %in% c("GSM1702177_7970368142_R05C02", "GSM1947213_6008581028_R01C01") ~ "Outlier",
                             !(Sample_name %in% c("GSM1702177_7970368142_R05C02", "GSM1947213_6008581028_R01C01")) ~ Outlier1)) %>% 
  dplyr::select(Sample, Sample_name, Trimester, Study, ArrayType, Outlier, X, Y, Sex=Fetal_Sex, predicted_sex, WrongSexInfo) %>% 
  dplyr::arrange(desc(WrongSexInfo)) %>% 
  write_csv(path = "/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_phenotypes_addEwastoolPredictedSex.csv")

```



### 2) `CheckGenotype` function

```{r checkGenotype}

CheckGenotype <- function(RGsetChara){
  registerDoParallel(cores=10)
  
  # get the object from character
  RGset <- get(RGsetChara)
  
  # get meta data
  phenoData <- pData(RGset)
  # phenoData to phenoData.dt
  phenoData.dt <- phenoData %>% as.data.frame() %>% rownames_to_column() %>% setDT()
  
  # read idat files and get beta
  sampleDirs <- phenoData$filenames
  meth = read_idats(idat_files = sampleDirs,quiet=FALSE)
  beta = dont_normalize(meth)

  # get detection P values
  detP <- minfi::detectionP(RGset)

  # get snps
  snps = meth$manifest[probe_type=='rs',index]
  snps = beta[snps,]
  genotypes = call_genotypes(snps,learn=TRUE)

  if(ncol(RGset)==1){
    if (!"outliers" %in% names(genotypes)) 
        stop("Invalid argument")
    log_odds = genotypes$outliers/(1 - genotypes$outliers)
    log_odds = mean(log2(log_odds), na.rm = TRUE)
    # log_odds
    phenoData.dt$log_odds = log_odds
  }else{
  
  phenoData.dt$log_odds = snp_outliers(genotypes)
  }
  
  table(phenoData.dt$log_odds > -4)

  phenoData.dt[phenoData.dt$log_odds > -4, ] %>% arrange(desc(log_odds))

  # assumed distribution of SNP-probe intensities. if probes are outoff distribution, there could be some technical variance (could to be used to indicate tissue contanmination)
  mxm_(genotypes)

  # plot genotype checking results
  par(mfrow=c(2,1))
  barplot(colMeans(detP), col=pal[6], las=2,
        cex.names=0.8, ylim=c(0,0.01),
        xlab = "",
        ylab = "Mean detection p-values",
        axisnames = FALSE)
  abline(h=0.01,col="red")

  barplot(phenoData.dt$log_odds,
        xlab = "placental tissue samples(n=53)",
        ylab = "Average log odds")
  abline(h=-4,col="red")

  # which(phenoDataNormal$Sample.ID%in%62)
  # ggplot(tplot, aes(x=V54))+geom_density()
  rm(meth)

  # return phenoData_withPredict, the df with predicted fetal sex
  return(list(detP=detP, phenoData.dt=as_tibble(phenoData.dt)))
}

```


### use `CheckGenotype`
```{r CheckGenotypeforAllTissues}

TissueRGsetNames <- c("GEO_RGset_450k_Tissue", "GEO_RGset_EPIC_Tissue")


for(i in TissueRGsetNames){
  
  assign(paste0("GEO_pd", str_replace(i, "GEO_RGset",""), "_ewastoolPurityCheck"),   
  CheckGenotype(RGsetChara = i))

}

# RGsetChara <- "GEO_RGset_450k_Decidua"

# save all phenoData with predicted fetal sex infotmation
# dfs <- paste0("GEO_pd", str_replace(TissueRGsetNames, "GEO_RGset",""), "_ewastoolPurityCheck") 
# GEO_phenotypes_ewastoolPurityCheck <- do.call(what=rbind, 
#                     args = mget(dfs))

GEO_phenotypes_ewastoolPurityCheck <- rbind(GEO_pd_450k_Tissue_ewastoolPurityCheck$phenoData.dt, GEO_pd_EPIC_Tissue_ewastoolPurityCheck$phenoData.dt)

detP450k <- GEO_pd_450k_Tissue_ewastoolPurityCheck$detP %>% as.data.frame() %>% rownames_to_column(var="Probe")

detEPIC <- GEO_pd_EPIC_Tissue_ewastoolPurityCheck$detP %>% as.data.frame() %>% rownames_to_column(var="Probe")

GEO_phenotypes_detP <- left_join(detP450k, detEPIC, by="Probe") %>% 
  column_to_rownames(var="Probe") %>%
  as.matrix() %>% 
  na.omit() %>% 
  colMeans() %>% 
  as.data.frame() %>% 
  rownames_to_column() %>% 
  `colnames<-`(c("Sample_name","detP"))


saveRDS(GEO_phenotypes_ewastoolPurityCheck,
  file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_phenotypes_ewastoolPurityCheck.rds")


GEO_phenotypes_ewastoolPurityCheck[GEO_phenotypes_ewastoolPurityCheck$log_odds> -4, ]


GEO_phenotypes_ewastoolPurityCheck_output <- GEO_phenotypes_ewastoolPurityCheck %>% 
  dplyr::select(Sample, Sample_name, Trimester, Additional_Info, Fetal_Sex, Study, ArrayType, log_odds) %>% 
  dplyr::mutate(Outlier1=ifelse(.$Additional_Info%in%"Outlier", "Outlier", "Pure")) %>% 
  dplyr::mutate(Outlier = case_when(Sample_name %in% c("GSM1702177_7970368142_R05C02", "GSM1947213_6008581028_R01C01") ~ "Outlier",
                             !(Sample_name %in% c("GSM1702177_7970368142_R05C02", "GSM1947213_6008581028_R01C01")) ~ Outlier1)) %>% 
  dplyr::select(Sample, Sample_name, Trimester, Sex=Fetal_Sex, Study, ArrayType, Outlier, log_odds) %>% 
  dplyr::left_join(GEO_phenotypes_detP, by="Sample_name") %>% 
  arrange(desc(log_odds))



GEO_phenotypes_ewastoolPurityCheck_output %>% 
  write_csv(path = "/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_phenotypes_ewastoolPurityCheck_output.csv")

```


### Draw plot for sample contanmination detected by ewastool

```{r}
colorFun <- colorRampPalette(brewer.pal(8,"Dark2")[1:2])

log_odd_plot <- GEO_phenotypes_ewastoolPurityCheck_output %>% 
  ggplot(aes(x=Sample_name, y=log_odds, colour=Outlier)) +
  geom_point()+
  geom_hline(yintercept = -4, colour="red")+
  labs(x="", y="log odds")+
  theme(axis.text.x = element_blank(),
        axis.ticks.x= element_blank(),
        axis.text.y = element_text(size=9, colour ="black"),
        axis.title=element_text(size=9, colour ="black"),#,face="bold"
        text = element_text(size=9, colour ="black"))

detP_plot <- GEO_phenotypes_ewastoolPurityCheck_output %>% 
  ggplot(aes(x=Sample_name, y=detP, colour=Outlier)) +
  geom_point()+
  geom_hline(yintercept = 0.01, colour="red")+
  labs(x="", y="P value")+
  theme(axis.text.x = element_blank(),
        axis.ticks.x= element_blank(),
        axis.text.y = element_text(size=9, colour ="black"),
        axis.title=element_text(size=9, colour ="black"),#,face="bold"
        text = element_text(size=9, colour ="black"))

ggpubr::ggarrange(detP_plot, log_odd_plot, ncol = 1, nrow = 2, common.legend = TRUE)



log_odd_placenta_plot <- GEO_phenotypes_ewastoolPurityCheck_output[GEO_phenotypes_ewastoolPurityCheck_output$Sample%in%"Placenta",] %>% 
  ggplot(aes(x=Sample_name, y=log_odds, colour=Outlier)) +
  geom_point()+
  geom_hline(yintercept = -4, colour="red")+
  labs(x="", y="log odds")+
  theme(axis.text.x = element_blank(),
        axis.ticks.x= element_blank(),
        axis.text.y = element_text(size=9, colour ="black"),
        axis.title=element_text(size=9, colour ="black"),#,face="bold"
        text = element_text(size=9, colour ="black"))

t <-  GEO_phenotypes_ewastoolPurityCheck_output[GEO_phenotypes_ewastoolPurityCheck_output$Sample%in%"Placenta",]
t[t$log_odds> -4,]

# par(mfrow=c(2,1))
#   barplot(colMeans(detP), col=pal[6], #las=2,
#         cex.names=0.8,
#         ylim=c(0,0.01),
#         xlab = "",
#         ylab = "Mean detection p-values",
#         axisnames = FALSE)
#   abline(h=0.01,col="red")
# 
#   barplot(GEO_phenotypes_ewastoolPurityCheck_output$log_odds,
#         xlab = "placental tissue samples(n=408)",
#         ylab = "Average log odds", col = colorFun(2)[factor(GEO_phenotypes_ewastoolPurityCheck_output$Outlier)])
#   abline(h=-4,col="red")
  
```







