---
title: "GEO data sets_ICR analysis"
author: "Qianhui"
date: "1/14/2019"
output: html_document
---

This document is mainly for part8 of the GEO data set analysis, that is match beta values with imprinting control regions (ICRs).


```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.align = "center", error = TRUE, echo = TRUE) # fig.width = 7, 
```


# 1. load data and packages
```{r package, warning=FALSE, message=FALSE}

library(tidyverse)
library(minfi)
library(magrittr)
library(tibble)
library(readxl)
library(doParallel)
library(Gviz)
library(dplyr)
library(reshape2)
library(minfi)
library(plyranges)
library(BSgenome.Hsapiens.UCSC.hg19)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(ggplot2)
library(ggpubr)
library(RColorBrewer)
pal <- brewer.pal(8, "Dark2")

# GEOmeta, GEO_phenotypes
load(file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_phenotypes_add10.RData")

# Join.df_v2, GEO_phenotypes_joinedSamples, MDS_joinedSamples
load(file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_joinedSamples551_Beta_MDS.RData")

# annotation needed for using functions
ann450k <- readRDS(file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/ann450k_GEO15sets.rds")
annEPIC <- readRDS(file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/annEPIC_GEO15sets.rds")

# Residuals, these residuals are Mvalues with control PCA1&2 removed
load(file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/allSamplesResiduals_ControlCorrected.rds")

# BM list after normalisation, but not corrected for batch across samples
NormBMlist <- readRDS(file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/NormBMlist.rds")


```


# Import ICRs

There are 59 ICRs in GRange object `ICRs_v1`.
```{r}

ICRs <- read_xlsx(path = "/home/qianhui/DNAme/Process_decidual/ICRs/ICR_DMR_excelFile_fromPervjakova_etal.xlsx")

# other ICRs are regions without probes in those regions
ICR_Type <- c(rep('Ubiquitously imprinted', 45), rep('PlacentaSpecific', 14), rep('otherICRs', 8))

ICRsGR <- ICRs %>% na.omit() %>% inset('ICR_Type', value=ICR_Type) %>% 
           mutate(Gene=`Gene locus`) %>% 
           mutate(seqnames=paste0('chr', .$Chr)) %>% 
           mutate(start=Start) %>% 
           mutate(end=Finish) %>% 
           mutate(MethylationOrigin=`Methylation origin`) %>% 
           mutate(MeanBetaValue=str_replace_all(`Infinium beta value (min-max)`, " [(].+[)]", "")) %>% 
           dplyr::select(ICR_Type:MeanBetaValue) %>% 
           dplyr::filter(ICR_Type!="otherICRs") %>% 
           as_granges()
  
```


# 2. Plot beta values and ICRs (1st trimester samples)

The placenta-specific ICR were ICR46-59; ICR 1-45 were ubiquitously imprinted ICRs.

## 2.1 GRange for all probes (T1)
```{r}

colorFun <- colorRampPalette(brewer.pal(8,"Dark2"))
registerDoParallel(cores = 5)

table(colnames(NormBMlist$NormBAll) %in% GEO_phenotypes$Sample_name)
Matched_pd <- GEO_phenotypes[match(colnames(NormBMlist$NormBAll), GEO_phenotypes$Sample_name),]

PhenodataPlacenta <- Matched_pd[Matched_pd$Sample%in%"Placenta" & Matched_pd$Trimester%in%"First",]
# PhenodataPlacenta <- Matched_pd[Matched_pd$Trimester%in%"First",]

PhenodataPlacenta$Outlier <- factor(ifelse(PhenodataPlacenta$Additional_Info%in%"Outlier", "Outlier", "Standard"))

# PhenodataPlacenta$Outlier <- ifelse(PhenodataPlacenta$Additional_Info%in%"Outlier", "Outlier", "Standard")
# PhenodataPlacenta$Sample2 <- 
#   ifelse(str_detect(PhenodataPlacenta$Sample, "lacent.+"), "Placenta", 
#          ifelse(str_detect(PhenodataPlacenta$Sample, "Maternal"), "Maternal whole blood", "Decidua"))
# PhenodataPlacenta$Sample2[which(PhenodataPlacenta$Outlier=="Outlier")] <- rep("Placental outliers",5)


Placenta_Mnorm_v1 <- NormBMlist$NormMAll[,colnames(NormBMlist$NormMAll)%in%PhenodataPlacenta$Sample_name]
Placenta_betaNorm <- NormBMlist$NormBAll[,colnames(NormBMlist$NormBAll)%in%PhenodataPlacenta$Sample_name]

# some preperation for plotting
## Get probe ranges and select probe not related with islands
Gset_placenta <- makeGenomicRatioSetFromMatrix(Placenta_betaNorm, array = "IlluminaHumanMethylation450k",
                                               annotation = "ilmn12.hg19",mergeManifest = TRUE,
                                               what = "Beta")

GRangeProbe <- Gset_placenta@rowRanges[, c(1:2,7)]

# add mcol to GRangeProbe
## get GRange of all probes: 200k
### add values to GRange object metadata columns
matchWithGRprobe <- base::match(GRangeProbe$Name, rownames(Placenta_betaNorm))
values(GRangeProbe) <- cbind(values(GRangeProbe), DataFrame(Placenta_betaNorm[matchWithGRprobe,], check.names = F)) 

### add seqinfo
bs_hg19 <- BSgenome.Hsapiens.UCSC.hg19
bs_hg19
### check names are the same or not
seqnames(seqinfo(bs_hg19))[1:24] == seqinfo(GRangeProbe) %>% seqnames()

### assign seqinfo to my GRange for all probes 626374
seqinfo(GRangeProbe) <- Seqinfo(seqnames = seqnames(seqinfo(bs_hg19))[1:24],
                                seqlengths = seqlengths(seqinfo(bs_hg19))[1:24],
                                isCircular = isCircular(seqinfo(bs_hg19))[1:24],
                                genome = genome(seqinfo(bs_hg19))[1:24])

## get annotation and remove probes related wtih CpG islands
# ann450ksub <- ann450k[match(rownames(Placenta_betaNorm), ann450k$Name), ] %>% as.data.frame()
GRangeProbesForICRs <- GRangeProbe

```

## 2.2 GRange for probes (Term)
```{r}
# for Term samples

PhenodataPlacenta_Term <- Matched_pd[Matched_pd$Sample%in%"Placenta" & Matched_pd$Trimester%in%"Term",]
# PhenodataPlacenta_Term <- Matched_pd[Matched_pd$Trimester%in%"Term",]

PhenodataPlacenta_Term$Outlier <- factor(ifelse(PhenodataPlacenta_Term$Additional_Info%in%"Outlier", "Outlier", "Standard"))

# PhenodataPlacenta_Term$Outlier <- ifelse(PhenodataPlacenta_Term$Additional_Info%in%"Outlier", "Outlier", "Standard")
# PhenodataPlacenta_Term$Sample2 <- PhenodataPlacenta_Term$Sample
# PhenodataPlacenta_Term$Sample2[which(PhenodataPlacenta_Term$Outlier=="Outlier")] <- rep("Placental outliers",4)

Placenta_Mnorm_v1_Term <- NormBMlist$NormMAll[,colnames(NormBMlist$NormMAll)%in%PhenodataPlacenta_Term$Sample_name]
Placenta_betaNorm_Term <- NormBMlist$NormBAll[,colnames(NormBMlist$NormBAll)%in%PhenodataPlacenta_Term$Sample_name]

# some preperation for plotting
## Get probe ranges and select probe not related with islands
Gset_placenta_Term <- makeGenomicRatioSetFromMatrix(Placenta_betaNorm_Term, 
                                                    array = "IlluminaHumanMethylation450k",
                                               annotation = "ilmn12.hg19",mergeManifest = TRUE,
                                               what = "Beta")

GRangeProbe_Term <- Gset_placenta_Term@rowRanges[, c(1:2,7)]

# add mcol to GRangeProbe
matchWithGRprobe_Term <- base::match(GRangeProbe_Term$Name, rownames(Placenta_betaNorm_Term))
values(GRangeProbe_Term) <- cbind(values(GRangeProbe_Term), DataFrame(Placenta_betaNorm_Term[matchWithGRprobe_Term,], 
                                                            check.names = F)) 

### add seqinfo
### check names are the same or not
seqnames(seqinfo(bs_hg19))[1:24] == seqinfo(GRangeProbe_Term) %>% seqnames()

### assign seqinfo to my GRange for all probes 626374
seqinfo(GRangeProbe_Term) <- Seqinfo(seqnames = seqnames(seqinfo(bs_hg19))[1:24],
                                seqlengths = seqlengths(seqinfo(bs_hg19))[1:24],
                                isCircular = isCircular(seqinfo(bs_hg19))[1:24],
                                genome = genome(seqinfo(bs_hg19))[1:24])

## get annotation and remove probes related wtih CpG islands
# ann450ksub <- ann450k[match(rownames(Placenta_betaNorm), ann450k$Name), ] %>% as.data.frame()
GRangeProbesForICRs_Term <- GRangeProbe_Term

```

## 2.3 Gviz plot for ICR (Chr9, GLIS3)
```{r}

# Gviz plot for each ICR
# 1:length(ICRsGR)
for(i in 46){

genome <- "hg19"

chrom <- as.character(seqnames(ICRsGR))[i]

start <- start(ICRsGR)[i]

end <- end(ICRsGR)[i]

# add extra space to plot (since plot range should be longer than this DMR)
minbase <- start - (0.01*(end-start))
maxbase <- end + (0.01*(end-start))

#### ideogram Track:
iTrack <- IdeogramTrack(genome = genome, chromosome = chrom, name = "", from=minbase, to=maxbase,
                        fontsize=9, fontcolor="black")

#### Genome Axis Track:
gTrack <- GenomeAxisTrack(col="black", cex=0.7, name = "", fontcolor="black", add53=T, 
                          # add35=T,
                          fontsize=8, col="black")
#### ICR track
ICRTrack <- AnnotationTrack(range=ICRsGR, gensome = genome, chromosome = chrom, 
                            start = start, end = end,
                            name = paste0("ICR\n(GLIS3)",i),
                            id="ICR of GLIS3",
                            fontcolor.feature="black",
                            col.title="black",
                            #groupAnnotation=NULL,
                            background.title="transparent",
                            # showId=TRUE,
                            showFeatureId=TRUE,
                            stacking = "dense")

#### create methylation data Track:
library(RColorBrewer)
# colorFun <- colorRampPalette(brewer.pal(9,"Blues")[3:9])
# colorFun <- colorRampPalette(brewer.pal(9,"YlOrRd"))

methTrack_T1 <- DataTrack(
  range = GRangeProbesForICRs[,4:35],
  genome = genome, chromosome = chrom, start = minbase, end = maxbase,
  # groups=factor(phenoDataPlacenta_GAordered$Gestational.Age),
  groups=factor(PhenodataPlacenta$Outlier),
  ylim= c(0, 1.00),
  # col=unique(colorFun(18)[factor(phenoDataPlacenta_GAordered$Gestational.Age)]),
  # col=brewer.pal(8,"Dark2")[c(3,4,2,1)],
  col=brewer.pal(8,"Dark2")[1:2],
  type = c("a", "p"), # a: line plot of average
  # type = "smooth",  baseline=0.5, span=1/50,
  name = "Beta value\n(First Trimester)",
  # background.panel = "black",
  fontsize.legend=9,
  # box.legend=TRUE,
  legend=FALSE,
  fontcolor.legend="black",
  background.title="transparent",
  col.title="black",
  col.axis="black",
  # cex.title=0.8,
  # cex.axis=0.8,
  cex.legend = 0.9
  )

methTrack_Term <- DataTrack(
  range = GRangeProbesForICRs_Term[,4:345],
  genome = genome, chromosome = chrom, start = minbase, end = maxbase,
  # groups=factor(phenoDataPlacenta_GAordered$Gestational.Age),
  groups=factor(PhenodataPlacenta_Term$Outlier),
  ylim= c(0, 1.00),
  # col=unique(colorFun(18)[factor(phenoDataPlacenta_GAordered$Gestational.Age)]),
  # col=brewer.pal(8,"Dark2")[c(3,4,2,1)],
  col=brewer.pal(8,"Dark2")[1:2],
  type = c("a", "p"), # a: line plot of average
  # type = "smooth",  baseline=0.5, span=1/50,
  name = "Beta value\n(Term)",
  # background.panel = "black",
  fontsize.legend=9,
  # box.legend=TRUE,
  fontcolor.legend="black",
  background.title="transparent",
  col.title="black",
  col.axis="black",
  # cex.title=0.8,
  # cex.axis=0.8,
  cex.legend = 0.9
  )

# tracks to draw
tracks <- list(iTrack, gTrack, ICRTrack, methTrack_T1, methTrack_Term)

#### set up relative size for each track:
sizes <- c(1,1,1,3,4)
plotTracks(tracks, 
           from = minbase-1 , to = maxbase+1, 
           showTitle=TRUE,
           add53 = TRUE, 
           # add35 = TRUE, 
           grid=FALSE,
           # lty.grid=3,
           sizes = sizes,
           #showBandId=TRUE,
           #showId=TRUE,
           #showFeatureId=TRUE,
           length(tracks))

tiff(filename = paste0("/home/qianhui/DNAme/Process_decidual/ICRs/ICRfigures/Gviz_T1samples_ICR_", i, "_v1.tiff"), 
     width = 10, height = 8, units = "cm", res = 300)
plotTracks(tracks, 
           from = minbase-1 , to = maxbase+1, 
           showTitle=TRUE,
           add53 = TRUE, 
           # add35 = TRUE, 
           grid=FALSE,
           # lty.grid=3,
           sizes = sizes,
           #showBandId=TRUE,
           #showId=TRUE,
           #showFeatureId=TRUE,
           length(tracks))
dev.off()
}

```

## 2.4 plot violin plot

<!-- ### average Beta in ICR (T1) -->
<!-- ```{r} -->
<!-- library(rlang) -->
<!-- ### your column names for your 96 ids, replace as necessary -->
<!-- nms <- colnames(values(GRangeProbesForICRs))[4:35] -->
<!-- # this will generate the mean calls -->
<!-- cols <- lapply(nms, function(.) { col <- ensym(.); quo(mean(!!col)) }) -->
<!-- # the column names for the summarize op -->
<!-- names(cols) <- paste0("mean_", nms) -->
<!-- # your code, the !!! unsplices the column names we have generated -->
<!-- ProbesBetaMeansInICR <- group_by_overlaps(ICRsGR, GRangeProbesForICRs) %>% summarise(!!! cols) -->

<!-- ``` -->

<!-- ### average Beta in ICR (Term) -->
<!-- ```{r} -->
<!-- library(rlang) -->
<!-- ### your column names for your 96 ids, replace as necessary -->
<!-- nms_term <- colnames(values(GRangeProbesForICRs_Term))[4:345] -->
<!-- # this will generate the mean calls -->
<!-- cols_term <- lapply(nms_term, function(.) { col <- ensym(.); quo(mean(!!col)) }) -->
<!-- # the column names for the summarize op -->
<!-- names(cols_term) <- paste0("mean_", nms_term) -->
<!-- # your code, the !!! unsplices the column names we have generated -->
<!-- ProbesBetaMeansInICR_term <- group_by_overlaps(ICRsGR, GRangeProbesForICRs_Term) %>% summarise(!!! cols_term) -->

<!-- ``` -->



```{r}
# T1 samples
ProbesBetaInICR <- find_overlaps(ICRsGR[ICRsGR$ICR_Type%in%"PlacentaSpecific",], GRangeProbesForICRs)[,c(6,8:39)] %>% 
  mcols() %>% unique() %>% as.data.frame() %>% 
  remove_rownames() #%>% column_to_rownames(var="Name")

# Term samples
ProbesBetaInICR_Term <- find_overlaps(ICRsGR[ICRsGR$ICR_Type%in%"PlacentaSpecific",], GRangeProbesForICRs_Term)[,c(6,8:349)] %>% 
  mcols() %>% unique() %>% as.data.frame() %>% 
  remove_rownames() #%>% column_to_rownames(var="Name")

# join T1 and Term samples
ProbesBetaInICR_T1Term <- ProbesBetaInICR %>% left_join(ProbesBetaInICR_Term, by="Name")

# ProbesBetaInICR_T1Term <- ProbesBetaMeansInICR %>% as.data.frame() %>% left_join(as.data.frame(ProbesBetaMeansInICR_term), by="query")

## change matrix to a long data frame
ProbesBetaInICR_long <- ProbesBetaInICR_T1Term %>% as.data.frame() %>% 
  gather(key = "sampleName",
         value = "Betas",
         # contains("mean_")
         -Name
         )

## add outlier information as a column
ProbesBetaInICR_long_v1 <- ProbesBetaInICR_long %>% mutate(sampleName= str_replace_all(sampleName, "^X", ""))
# ProbesBetaInICR_long_v1 <- ProbesBetaInICR_long %>% mutate(sampleName= str_replace_all(sampleName, "mean_", ""))

PhenodataPlacentaT1Term <- rbind(PhenodataPlacenta, PhenodataPlacenta_Term)
PhenodataPlacentaT1Term_v1 <- PhenodataPlacentaT1Term %>% mutate(sampleName=Sample_name) %>% 
  dplyr::select(sampleName, Outlier, Trimester)

ProbesBetaInICR_long_v2 <- left_join(ProbesBetaInICR_long_v1, PhenodataPlacentaT1Term_v1, by="sampleName")


#color setting
pal <- brewer.pal(8, "Dark2")
pal2 <- c("#08519C",pal[2])
namedPal2 <- pal2 %>% `names<-`(unique(ProbesBetaInICR_long_v2$Outliers))

# plot and save the violin plot
tiff(file = "/home/qianhui/DNAme/Process_decidual/ICRs/ICRfigures/ICRprobes_violinPlot_374Placenta.tiff", 
     width = 10, height = 7, units = "cm", res = 300)
theme_set(theme_pubr())
ggplot(ProbesBetaInICR_long_v2, aes(Outlier, Betas, fill=Outliers))+ 
  geom_violin(trim = T, aes(fill=Outlier))+ 
  scale_fill_brewer(palette = "Dark2", labels=c("Outlier"="Outliers", "Standard"="Standard"))+
  labs(x="", y="Beta values", fill="Placenta")+
  geom_boxplot(width=0.1, fill="white")+
  facet_wrap(~Trimester)+
  theme(text = element_text(size=10, colour = "black"),
        # axis.text.x=element_text(size=6, colour = "black"),
        axis.text.x=element_blank(),
        axis.text.y=element_text(size=10, colour = "black"),
        legend.position = "right", 
        strip.text.x = element_text(size = 10, colour = "black"))
dev.off()

```


# 3. Plot beta values and ICRs (Term samples)

The placenta-specific ICR were ICR46-59; ICR 1-45 were ubiquitously imprinted ICRs.

```{r eval=FALSE}

# colorFun <- colorRampPalette(brewer.pal(8,"Dark2"))
# registerDoParallel(cores = 5)
# 
# table(colnames(NormBMlist$NormBAll) %in% GEO_phenotypes$Sample_name)
# Matched_pd <- GEO_phenotypes[match(colnames(NormBMlist$NormBAll), GEO_phenotypes$Sample_name),]
# 
# # PhenodataPlacenta <- Matched_pd[Matched_pd$Sample%in%"Placenta" & Matched_pd$Trimester%in%"Term",]
# PhenodataPlacenta <- Matched_pd[Matched_pd$Trimester%in%"Term",]
# 
# PhenodataPlacenta$Outlier <- ifelse(PhenodataPlacenta$Additional_Info%in%"Outlier", "Outlier", "Standard")
# PhenodataPlacenta$Sample2 <- PhenodataPlacenta$Sample
# PhenodataPlacenta$Sample2[which(PhenodataPlacenta$Outlier=="Outlier")] <- rep("Placental outliers",4)
# 
# 
# Placenta_Mnorm_v1 <- NormBMlist$NormMAll[,colnames(NormBMlist$NormMAll)%in%PhenodataPlacenta$Sample_name]
# Placenta_betaNorm <- NormBMlist$NormBAll[,colnames(NormBMlist$NormBAll)%in%PhenodataPlacenta$Sample_name]
# 
# # some preperation for plotting
# ## Get probe ranges and select probe not related with islands
# Gset_placenta <- makeGenomicRatioSetFromMatrix(Placenta_betaNorm, array = "IlluminaHumanMethylation450k",
#                                                annotation = "ilmn12.hg19",mergeManifest = TRUE,
#                                                what = "Beta")
# 
# GRangeProbe <- Gset_placenta@rowRanges[, c(1:2,7)]
# 
# # add mcol to GRangeProbe
# ## get GRange of all probes: 200k
# ### add values to GRange object metadata columns
# matchWithGRprobe <- base::match(GRangeProbe$Name, rownames(Placenta_betaNorm))
# values(GRangeProbe) <- cbind(values(GRangeProbe), DataFrame(Placenta_betaNorm[matchWithGRprobe,], check.names = F)) 
# 
# ### add seqinfo
# bs_hg19 <- BSgenome.Hsapiens.UCSC.hg19
# bs_hg19
# ### check names are the same or not
# seqnames(seqinfo(bs_hg19))[1:24] == seqinfo(GRangeProbe) %>% seqnames()
# 
# ### assign seqinfo to my GRange for all probes 626374
# seqinfo(GRangeProbe) <- Seqinfo(seqnames = seqnames(seqinfo(bs_hg19))[1:24],
#                                 seqlengths = seqlengths(seqinfo(bs_hg19))[1:24],
#                                 isCircular = isCircular(seqinfo(bs_hg19))[1:24],
#                                 genome = genome(seqinfo(bs_hg19))[1:24])
# 
# ## get annotation and remove probes related wtih CpG islands
# # ann450ksub <- ann450k[match(rownames(Placenta_betaNorm), ann450k$Name), ] %>% as.data.frame()
# GRangeProbesForICRs <- GRangeProbe


# Gviz plot for each ICR
# 1:length(ICRsGR)
for(i in 46){

genome <- "hg19"

chrom <- as.character(seqnames(ICRsGR))[i]

start <- start(ICRsGR)[i]

end <- end(ICRsGR)[i]

# add extra space to plot (since plot range should be longer than this DMR)
minbase <- start - (0.1*(end-start))
maxbase <- end + (0.1*(end-start))

#### ideogram Track:
iTrack <- IdeogramTrack(genome = genome, chromosome = chrom, name = "", from=minbase, to=maxbase,
                        fontsize=9, fontcolor="black")

#### Genome Axis Track:
gTrack <- GenomeAxisTrack(col="black", cex=0.9, name = "", fontcolor="black", add53=T, add35=T,
                          fontsize=9, col="black")
#### ICR track
ICRTrack <- AnnotationTrack(range=ICRsGR, gensome = genome, chromosome = chrom, 
                            start = start, end = end,
                            name = paste0("Published ICR ",i),
                            id="ICR",
                            fontcolor.feature="black",
                            col.title="black",
                            #groupAnnotation=NULL,
                            background.title="transparent",
                            showId=FALSE,
                            showFeatureId=FALSE,
                            stacking = "dense")

#### create methylation data Track:
library(RColorBrewer)
# colorFun <- colorRampPalette(brewer.pal(9,"Blues")[3:9])
# colorFun <- colorRampPalette(brewer.pal(9,"YlOrRd"))

methTrack <- DataTrack(
  range = GRangeProbesForICRs[,4:361],
  genome = genome, chromosome = chrom, start = minbase, end = maxbase,
  # groups=factor(phenoDataPlacenta_GAordered$Gestational.Age),
  groups=factor(PhenodataPlacenta$Sample2),
  ylim= c(0, 1.00),
  # col=unique(colorFun(18)[factor(phenoDataPlacenta_GAordered$Gestational.Age)]),
  col=c("#000000", "#BD0026", "#984EA3", "#D95F02", "#1B9E77", "#999999"),
  type = c("a", "p"), # a: line plot of average
  # type = "smooth",  baseline=0.5, span=1/50,
  name = "DNA Methylation\n(beta value)",
  # background.panel = "black",
  fontsize.legend=8,
  # box.legend=TRUE,
  fontcolor.legend="black",
  lineheight.legend=2,
  background.title="transparent",
  col.title="black",
  col.axis="black",
  # cex.title=0.8,
  # cex.axis=0.8,
  cex.legend = 0.5
  )

# tracks to draw
tracks <- list(iTrack, gTrack, ICRTrack, methTrack)

#### set up relative size for each track:
sizes <- c(2,2,2,4)
plotTracks(tracks, 
           from = minbase-1 , to = maxbase+1, 
           showTitle=TRUE,
           add53 = TRUE, add35 = TRUE, 
           grid=FALSE,
           # lty.grid=3,
           sizes = sizes,
           #showBandId=TRUE,
           #showId=TRUE,
           #showFeatureId=TRUE,
           length(tracks))

tiff(filename = paste0("/home/qianhui/DNAme/Process_decidual/ICRs/ICRfigures/Gviz_TermSamples_ICR_", i, ".tiff"), 
     width = 8, height = 6.5, units = "cm", res = 300)
plotTracks(tracks, 
           from = minbase-1 , to = maxbase+1, 
           showTitle=TRUE,
           add53 = TRUE, add35 = TRUE, 
           grid=FALSE,
           # lty.grid=3,
           sizes = sizes,
           #showBandId=TRUE,
           #showId=TRUE,
           #showFeatureId=TRUE,
           length(tracks))

dev.off()
}

```