---
title: "technical report (mainly 450k data, 15 GEO data sets used)"
author: "Qianhui"
date: "21/11/2018"
output: html_document
---

This document is mainly for part5 of the GEO data set analysis, that is the quality control (check fetal sex) of data sets from different tissue types.

I used methods from `minfi` package to identify samples with fetal sex misslabelled (check fetal sex)

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.align = "center", error = TRUE, echo = TRUE) # fig.width = 7, 
```


# 1. load data and packages
```{r package, warning=FALSE, message=FALSE}

library(tidyverse)
library(minfi)
library(magrittr)
library(tibble)
library(data.table)
library(limma)
# library(ewastools)
library(ENmix)
library(FactoMineR)
library(doParallel)
library(ggplot2)
library(RColorBrewer)
pal <- brewer.pal(8, "Dark2")

# GEOmeta, GEO_phenotypes
load(file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/GEOphenoDataAll.RData")

```

```{r}
GEO_baseDir <- file.path("/home/qianhui/DNAme/Process_decidual/GEO_dataSets/IDAT_all")
  # get 450k meta data
  GEO_phenotypes_EPIC <- GEO_phenotypes_Filtered[GEO_phenotypes_Filtered$ArrayType=='EPIC',]
  # get 450k, meta data for the tissue
  GEO_phenotypes_EPIC_Tissue <- GEO_phenotypes_EPIC %>% 
  filter(str_detect(Sample, pattern=paste0("^", Tissue, "$"))) %>% 
  filter(str_detect(Disease, pattern="Uncomplicated$"))
  # get the RGset for this tissue
  GEO_RGset_EPIC_Tissue <- read.metharray.exp(base = GEO_baseDir, targets = GEO_phenotypes_EPIC_Tissue)
  # GEO_phenoData_450k_Tissue <- pData(GEO_RGset_450k_Tissue) #chain data & phenotypes together
  
```

# 3. ewastools: check sex and contanmincation of sampels.

## Functions to use ewastools to check fetal sex and genotypes.

### 1) `CheckFetalSex` Function
```{r checkFetalSex}
# let argument RGsetChara to be a charactor

CheckFetalSex <- function(RGsetChara){
  registerDoParallel(cores=10)
  # get the object from character
  RGset <- get(RGsetChara)
  # get meta data
  phenoData <- pData(RGset)
  
  # get Genomic methylSets
  ctrls <- getProbeInfo(RGset,type="Control")
  ctrls <- ctrls[ctrls$Address %in% featureNames(RGset),]
  ctrl_r <- getRed(RGset)[ctrls$Address,]
  ctrl_g <- getGreen(RGset)[ctrls$Address,]
  CG.controls <- ctrls$Type %in% c("NORM_C","NORM_G")
  AT.controls <- ctrls$Type %in% c("NORM_A","NORM_T")
  if(ncol(RGset)==1){ # can't use relic if there is only one sample in the group
      # cg_grn <- ctrl_g[CG.controls] %>% as.matrix()
      # rownames(cg_grn) = ctrls$ExtendedType[CG.controls]
      # at_red <- ctrl_r[AT.controls] %>% as.matrix()
      # rownames(at_red) = ctrls$ExtendedType[AT.controls]
      # GMethSet <- RGset %>% preprocessRaw() %>% mapToGenome()
    print("Only one sample in this group, can't predict sex")
    }else{
      cg_grn <- ctrl_g[CG.controls,]
      rownames(cg_grn) = ctrls$ExtendedType[CG.controls]
      at_red <- ctrl_r[AT.controls,]
      rownames(at_red) = ctrls$ExtendedType[AT.controls]
      GMethSet <- RGset %>% preprocessRaw() %>% relic(at_red=at_red, cg_grn=cg_grn) %>% mapToGenome()
    }
    
  
  # sex check
  # phenoData to phenoData.dt
  phenoData.dt <- phenoData %>% as.data.frame() %>% rownames_to_column() %>% setDT()
  
  # add average total beta values for probes on X and Y chr
  estSex <- getSex(GMethSet)
  GMsetEx <- addSex(GMsetEx, sex = estSex)
  
  # predict fetal sex
  phenoData.dt$predicted_sex = predict_sex(phenoData.dt$X,phenoData.dt$Y
                           # which(phenoData.dt$Fetal_Sex%in%'Male'),
                           # which(phenoData.dt$Fetal_Sex%in%'Female')
                           )

  phenoData.dt_v1 <- phenoData.dt %>%
    mutate(predicted_sex = str_replace_all(predicted_sex, c("m"="Male","f" ="Female")))
  
  phenoData.dt_v2 <- setDT(phenoData.dt_v1)
  
  # plot the X/Y chr intensities
  # it is better to put pch into 3 groups (red color marked the misslabelled samples)
  plot(Y ~ X,data=phenoData.dt_v2,
       pch=ifelse(phenoData.dt_v2$Fetal_Sex=="Female",1,
           ifelse(phenoData.dt_v2$Fetal_Sex=="Male", 4,13)),
       asp=1,xlab="Normalized X chromosome intensities",ylab="Normalized Y chromosome intensities")
  points(Y ~ X,data=phenoData.dt_v2[Fetal_Sex !=predicted_sex],
         pch=ifelse(phenoData.dt_v2[Fetal_Sex!=predicted_sex]$Fetal_Sex=="Female",1,
             ifelse(phenoData.dt_v2[Fetal_Sex!=predicted_sex]$Fetal_Sex=="Male",4,13)),
         col=2)
  legend("topright",pch=c(1,4,13),legend=c("Female","Male", "NA"))
  title(main = RGsetChara)
  
  # samples with fetal sex in meta data not same as predicted fetal sex
  phenoData.dt_v2[Fetal_Sex !=predicted_sex] %>% as_tibble %>% 
    dplyr::select(rowname:Study, X:predicted_sex) %>% print(n=Inf)
  
  # remove intermediat data
  rm(meth)
  
  # return phenoData_withPredict, the df with predicted fetal sex
  return(as_tibble(phenoData.dt_v2))
}

```


### use `CheckFetalSex`
```{r CheckFetalSexforAllTissues}

TissueRGsetNames <- c("GEO_RGset_450k_normalPlacenta", "GEO_RGset_EPIC_normalPlacenta",
                      "GEO_RGset_450k_Amnion", 
                      "GEO_RGset_450k_Chorion", 
                      "GEO_RGset_450k_Decidua", "GEO_RGset_EPIC_Decidua",
                      "GEO_RGset_450k_UC", "GEO_RGset_450k_UCB"
                      )


for(i in TissueRGsetNames){
  # check fetal sex for each subset data sets
  # and save the output df, the phenodata with predicted fetal sex
  assign(paste0("GEO_phenoData", str_replace(i, "GEO_RGset",""), "_PreFS"),   
         CheckFetalSex(RGsetChara = i))

}

# save all phenoData with predicted fetal sex infotmation
save(GEO_phenoData_450k_normalPlacenta_PreFS, GEO_phenoData_EPIC_normalPlacenta_PreFS,
     GEO_phenoData_450k_Amnion_PreFS, GEO_phenoData_EPIC_Amnion_PreFS,
     GEO_phenoData_450k_Chorion_PreFS, GEO_phenoData_EPIC_Chorion_PreFS,
     GEO_phenoData_450k_Decidua_PreFS, GEO_phenoData_EPIC_Decidua_PreFS,
     GEO_phenoData_450k_MaternalBlood_PreFS, 
     GEO_phenoData_450k_mesenchyme_PreFS,
     GEO_phenoData_450k_trophoblast_PreFS, 
     GEO_phenoData_450k_UC_PreFS, 
     GEO_phenoData_450k_UCB_PreFS,
  file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_allPhenoData_addPredictedFetalSex.RData"
)

RGsetChara="GEO_RGset_450k_Chorion"

```



