---
title: "technical report (mainly 450k data, 15 GEO data sets)"
author: "Qianhui"
date: "28/08/2018"
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.align = "center", error = TRUE, echo = TRUE) # fig.width = 7, 
```


# 1.Packages and data required

```{r package, warning=FALSE, message=FALSE}
# library(TxDb.Hsapiens.UCSC.hg19.knownGene)
# library(BSgenome.Hsapiens.UCSC.hg19)

library(dplyr)
library(tibble)
library(readr)
library(magrittr)
library(data.table)
library(tibble)
library(stringi)
library(tidyverse)
library(reshape2)
library(stringr)
library(minfi)
library(IlluminaHumanMethylation450kmanifest)
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
library(sva)
library(RPMM)
library(matrixStats)
library(wateRmelon)
library(limma)
library(missMethyl)
# library(ewastools)
library(ENmix)
library(ChAMP)
library(FactoMineR)
library(plyranges)

library(GenomicRanges)
library(RColorBrewer)
library(ggplot2)
library(plotrix)
library(gridExtra)
library(doParallel)

library(GEOquery)
# colours to use
library(RColorBrewer)
pal <- brewer.pal(8, "Dark2")


colorFun <- colorRampPalette(brewer.pal(9, "Set1"))
pal2 <- colorFun(10)

theme_set(theme_bw())

# crossreactiveProbes <- read_csv("/home/qianhui/DNAme/Process-QH/13059_2016_1066_MOESM1_ESM-cross-reactive.csv")
# colnames(crossreactiveProbes)[1] <- "ProbeNames"

# Placenta_betaNorm <- readRDS(file = "/home/qianhui/DNAme/Process_placenta_workflow/Placenta_betaBMIQ617284_5rm.rds")
# Placenta_Mnorm_v1 <- readRDS(file = "/home/qianhui/DNAme/Process_placenta_workflow/Placenta_Mnorm617284_5rm.rds")

# phenoDataPlacenta91 <- readRDS(file = "/home/qianhui/DNAme/Process_placenta_workflow/phenoDataPlacenta91.rds")
# phenoDataPlacenta <- readRDS(file = "/home/qianhui/DNAme/Process_placenta_workflow/phenoDataPlacenta.rds")
# 
# Placenta_betaNorm <- readRDS(file = "/home/qianhui/DNAme/Process_placenta_workflow/Placenta_betaBMIQ_96.rds")
# Placenta_Mnorm_v1 <- readRDS(file = "/home/qianhui/DNAme/Process_placenta_workflow/Placenta_Mnorm_96.rds")

# raw data after filtering and before bg correction and normalisation
# methPlacenta_PBeadXsnpXY <- readRDS(file = "/home/qianhui/DNAme/Process_placenta_workflow/placenta_MethylSetRaw_drop_PBeadXsnpXY.rds")

# data after bg correction
# methylsetPlacenta_bgCorrENmix <- readRDS(file = "/home/qianhui/DNAme/Process_placenta_workflow/placenta_methylsetPlacenta_bgCorrENmix.rds")

```


# 2. get phenodata/metadata for all GEO data sets (15)
'GSE41933'--unbilical cord; 'GSE66459'--unbilical cord blood

```{r meta, eval=FALSE}
# Download of the meta data for each data set
# GSE73375 <- getGEO('GSE73375', destdir  = '/home/qianhui/DNAme/GAclock/RepeatBensClock/GEO_data/')

GEO_num <- c('GSE66210', 'GSE74738', 'GSE69502', 'GSE75196', 
             'GSE75248', 'GSE93208', 'GSE115508', 'GSE120250', 
             'GSE108567', 'GSE100197', 'GSE98224', 'GSE106089', 
             'GSE71678', 'GSE113600', 'GSE66459')

#GSE93429, GSE79262, GSE41933 removed because they have no raw IDAT files.
#GSE108567 removed because it's a repeated data set for reanalysis purpose.

library(doParallel)
registerDoParallel(cores = 10)

for(i in 1:15){
  assign(paste0(GEO_num[i]),
         getGEO(GEO_num[i], destdir  = '/home/qianhui/DNAme/Process_decidual/GEO_dataSets/MetaData_GEO15'))
}

# Extract the meta data
# GSE73375.meta <- pData(phenoData(GSE73375[[1]]))

for(i in 1:15){
  # if data sets contain data form differenct platforms
  if(length(get(GEO_num[i]))>1){
    
  DNAmePlatform <- str_detect(names(get(GEO_num[i])), pattern = 'GPL13534')
  assign(paste0(GEO_num[i], ".meta"),
         pData(phenoData(get(GEO_num[i])[DNAmePlatform][[1]])))
  
  }else{
  # if data only from one platform, we can extract meta data directly 
  assign(paste0(GEO_num[i], ".meta"),
         pData(phenoData(get(GEO_num[i])[[1]])))
  } # end if
} # end for

# Save all the meta data into one file
save(GSE66210.meta, GSE74738.meta, GSE69502.meta, GSE75196.meta, GSE75248.meta, 
     GSE93208.meta, GSE115508.meta, GSE120250.meta, GSE108567.meta, GSE100197.meta, 
     GSE98224.meta, GSE106089.meta, GSE71678.meta, GSE113600.meta, GSE66459.meta,
     file='/home/qianhui/DNAme/Process_decidual/RDSfiles/pdataOrMetadata_GEO_15datasets.Rdata' )

```


# 3. get raw data
save raw data to the baseDir

```{r, eval=FALSE}

registerDoParallel(cores = 10)

for(i in 1:15){
  # assign(paste0(GEO_num[i], ".raw"),
  getGEOSuppFiles(GEO_num[i], 
                  # makeDirectory = FALSE,
                  # filter_regex='RAW',
                  baseDir = '/home/qianhui/DNAme/Process_decidual/GEO_dataSets')
}

```


# 4. untar and unzip raw data (mainly IDAT files)
First untar .tar files in bash, then unzip files in R
```{r, eval=FALSE}

GEO_num %>% head()

write.table(GEO_num, file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_GSEids.txt", 
            quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

for(n in 1:15){
# GEOid <- GEO_num[n]
  dir <- '/home/qianhui/DNAme/Process_decidual/GEO_dataSets/'
  gzFileNames <- list.files(paste0('/home/qianhui/DNAme/Process_decidual/GEO_dataSets/',
                                   GEO_num[n]), pattern = ".gz")
# some folder have more than one gz files, so I used if else function
if(length(gzFileNames)==0){
  print("no gz files, no need to use gunzip")

  }else{

  for(gz in gzFileNames){
  gunzip(filename = paste0(dir, GEO_num[n], "/", gz), overwrite = TRUE)
  } # end for gz
 } # end if
} # end for n

```
ðŸ˜€


# 5.reorganise meta data for all GEO data sets

Read the data to RG channel Set using `minfi` package.

## 1. GSE66210 from GEO database. `GEO_num[1]`
```{r ReadData1}
load(file = '/home/qianhui/DNAme/Process_decidual/RDSfiles/pdataOrMetadata_GEO_15datasets.Rdata')

registerDoParallel(cores = 10)

# establish a proper samplesheet for 
GSE66210.meta_v1 <- GSE66210.meta %>% 
  mutate(GEO_accession=geo_accession) %>% 
  mutate(Sample=str_replace_all(`tissue:ch1`, c('maternal whole blood'= 'Maternal whole blood', 'chorionic villus'='Placenta'))) %>% 
  mutate(Sample_name = str_replace_all(supplementary_file, c("_(Red|Grn).idat.gz"="", ".+/"=""))) %>% 
  mutate(Sample_nameToRowname=Sample_name) %>% 
  mutate(Disease = str_replace_all(`pregnancy:ch1`, 'normal pregnancy', 'Uncomplicated')) %>% 
  inset('Gestation', value=rep('NA', nrow(GSE66210.meta))) %>% 
  inset('Trimester', value=rep('First', nrow(GSE66210.meta))) %>% 
  inset('Additional_Info', value=rep('NA', nrow(GSE66210.meta))) %>% 
  inset('Additional_Info_2', value=rep('NA', nrow(GSE66210.meta))) %>% 
  mutate(Fetal_Sex=str_replace_all(`individual:ch1`, c('female fetus'='Female', 'male fetus'='Male'))) %>% 
  inset('ArrayType', value=rep('450K', nrow(GSE66210.meta))) %>% 
  inset('Study', value=rep('GSE66210', nrow(GSE66210.meta))) %>% 
  mutate('Array'=str_replace_all(description.1, "Sa.+_", "")) %>% 
  mutate('Slide'=str_replace_all(description.1, c("Sa.+ "="", "_.+"=""))) %>% 
  dplyr::select(GEO_accession:Slide) %>% 
  column_to_rownames(var='Sample_nameToRowname')

write_csv(GSE66210.meta_v1, path = "/home/qianhui/DNAme/Process_decidual/GEO_dataSets/GSE66210/GSE66210.meta_v1.csv")

##getwd()
# write function for get RG sets
getRGset <- function(GEOid){
  
baseDir <- file.path(paste0("/home/qianhui/DNAme/Process_decidual/GEO_dataSets/",GEOid))
phenotypes <- read.metharray.sheet(baseDir, pattern = "_v1.csv$")
RGset450k <- read.metharray.exp(base = baseDir, targets = phenotypes)
phenoData <- pData(RGset450k) #chain data & phenotypes together

saveRDS(phenoData, file = paste0("/home/qianhui/DNAme/Process_decidual/RDSfiles/", GEOid, "phenoData.rds"))

return(RGset450k)
}

ann450k <- getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)
saveRDS(ann450k, file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/ann450k_GEO15sets.rds")

library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
annEPIC <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)

saveRDS(annEPIC, file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/annEPIC_GEO15sets.rds")
# get RGset for GSE66210
# GSE66210.RGset <- getRGset('GSE66210')

```


## `GEO_num[2]` from GEO database. 
```{r ReadData2}
registerDoParallel(cores = 10)

# establish a proper samplesheet for 
GA <- GSE74738.meta$`fetal gestational age (weeks)/trimester:ch1`
GA_trimester <- str_replace_all(GA, c('na'='NA', '1st_trimester'='First', '2nd_trimester'='Second', '3rd_trimester'='Third',
                                      '<10w'='10', '12w 35-38d'='12', '15w by dates 9w size 53d'='15',
                                      '8w 37d DA'='8', '8w 48 days DA'='8', '9w 26-30d DA'='9'))
FSnames <- GSE74738.meta$`fetal sex/sex:ch1` %>% table() %>% names()
GSE74738.meta$`fetal sex/sex:ch1`[GSE74738.meta$`fetal sex/sex:ch1`==FSnames[7]] <- "M"

GSE74738.meta_v1 <- GSE74738.meta %>% 
  mutate(GEO_accession=geo_accession) %>% 
  mutate(Sample=str_replace_all(`sample tissue:ch1`, 
         c('amnioniotic membrane'='Amnion',
           'amniotic membrane'='Amnion',
           'chorionic membrane'='Chorion',
         'cord blood'='Umbilical cord blood', 'placental chorionic villi'='Placenta', 'placental decidua'='Decidua', 
         'whole maternal blood'='Maternal whole blood'))) %>% 
  mutate(Sample_name = str_replace_all(supplementary_file, c("_(Red|Grn).idat.gz"="", ".+/"=""))) %>% 
  mutate(Sample_nameToRowname=Sample_name) %>% 
  mutate(Disease = str_replace_all(`status/group:ch1`, c('Ab. MSS '= 'Ab.MSS', 'Chromosomically Normal Miscarriage'='Miscarriage',
                                                         'control'='Uncomplicated'))) %>% 
  mutate(Gestation= str_replace_all(GA, c('na'='NA', '1st_trimester'='NA', '2nd_trimester'='NA', '3rd_trimester'='NA',
                                          '<10w'='10', '12w 35-38d'='12', '15w by dates 9w size 53d'='15',
                                          '8w 37d DA'='8', '8w 48 days DA'='8', '9w 26-30d DA'='9'))) %>% 
  inset('Trimester', value=ifelse(GA_trimester=='NA', 'NA', 
                                  ifelse(GA_trimester=='First', 'First', 
                                  ifelse(GA_trimester=='Second', 'Second', 
                                  ifelse(GA_trimester=='Third', 'Third', 
                                  ifelse(as.numeric(GA_trimester)>=37, 'Term', 'Third')))))) %>% 
  inset('Additional_Info', value=rep('NA', nrow(GSE74738.meta))) %>% 
  inset('Additional_Info_2', value=rep('NA', nrow(GSE74738.meta))) %>% 
  mutate(Fetal_Sex= str_replace_all(`fetal sex/sex:ch1`, c('F'='Female', 'M'='Male', ' '='', 'pool_na'='NA'))) %>% 
  inset('ArrayType', value=rep('450K', nrow(GSE74738.meta))) %>% 
  inset('Study', value=rep('GSE74738', nrow(GSE74738.meta))) %>% 
  mutate('Array'= `450k sentrix_position:ch1`) %>% 
  mutate('Slide'= `450k sentrix_id:ch1`) %>% 
  dplyr::select(GEO_accession:Slide) %>% 
  column_to_rownames(var='Sample_nameToRowname')

ControlSamples <- GSE74738.meta_v1 %>% filter(str_detect(Sample, "Placenta") & Disease=='Uncomplicated')
ControlSamples$Gestation[ControlSamples$Gestation!='NA'] %>% range()

write_csv(GSE74738.meta_v1, path = "/home/qianhui/DNAme/Process_decidual/GEO_dataSets/GSE74738/GSE74738.meta_v1.csv")

##getwd()
# get RGset for GSE66210
# GSE74738.RGset <- getRGset('GSE74738')

```


## `GEO_num[3]` from GEO database. 
```{r ReadData3}
registerDoParallel(cores = 10)

# establish a proper samplesheet for 
GSE69502.meta_v1 <- GSE69502.meta %>% 
  mutate(GEO_accession=geo_accession) %>% 
  mutate(Sample= str_replace_all(`sample tissue:ch1`, 'chorionic villi', 'Placenta')) %>% 
  mutate(Sample_name = str_replace_all(supplementary_file, c("_(Red|Grn).idat.gz"="", ".+/"=""))) %>% 
  mutate(Sample_nameToRowname=Sample_name) %>% 
  mutate(Disease = str_replace_all(`ntd status:ch1`, 'control', 'Uncomplicated')) %>% 
  mutate(Gestation =`fetal gestational age (weeks):ch1`) %>% 
  inset('Trimester', value=rep('Second', nrow(GSE69502.meta))) %>% 
  inset('Additional_Info', value=rep('NA', nrow(GSE69502.meta))) %>% 
  inset('Additional_Info_2', value=rep('NA', nrow(GSE69502.meta))) %>% 
  mutate(Fetal_Sex= str_replace_all(`fetal sex:ch1`, c('F'='Female', 'M'='Male'))) %>% 
  inset('ArrayType', value=rep('450K', nrow(GSE69502.meta))) %>% 
  inset('Study', value=rep('GSE69502', nrow(GSE69502.meta))) %>% 
  mutate(Array= str_replace_all(description.3, ".+ ", "")) %>% 
  mutate(Slide= str_replace_all(description.2, ".+ ", "")) %>% 
  dplyr::select(GEO_accession:Slide) %>% 
  column_to_rownames(var='Sample_nameToRowname')

write_csv(GSE69502.meta_v1, path = "/home/qianhui/DNAme/Process_decidual/GEO_dataSets/GSE69502/GSE69502.meta_v1.csv")

##getwd()
# get RGset for GSE66210
# GSE69502.RGset <- getRGset('GSE69502')

```


## `GEO_num[4]` from GEO database. 
```{r ReadData4}
registerDoParallel(cores = 10)

# establish a proper samplesheet for 
GSE75196.meta_v1 <- GSE75196.meta %>% 
  mutate(GEO_accession=geo_accession) %>% 
  mutate(Sample=str_replace_all(`tissue:ch1`, 'placenta', 'Placenta')) %>% 
  mutate(Sample_name = str_replace_all(supplementary_file, c("_(Red|Grn).idat.gz"="", ".+/"=""))) %>% 
  mutate(Sample_nameToRowname=Sample_name) %>% 
  mutate(Disease = str_replace_all(`disease:ch1`, c('normal/healthy'='Uncomplicated', 'preeclampsia'='PE'))) %>% 
  mutate(Gestation =`gestation (wk):ch1`) %>% 
  inset('Trimester', value=ifelse(GSE75196.meta$`gestation (wk):ch1` >= 37, 'Term', 'Third')) %>% 
  inset('Additional_Info', value=rep('NA', nrow(GSE75196.meta))) %>% 
  inset('Additional_Info_2', value=rep('NA', nrow(GSE75196.meta))) %>% 
  mutate(Fetal_Sex= `Sex:ch1`) %>% 
  inset('ArrayType', value=rep('450K', nrow(GSE75196.meta))) %>% 
  inset('Study', value=rep('GSE75196', nrow(GSE75196.meta))) %>% 
  mutate(Array= str_extract(Sample_name, pattern = "R..C..")) %>% 
  mutate(Slide=  str_extract(Sample_name, pattern= "_(.*?)_")) %>% 
  mutate(Slide=  str_replace_all(Slide, '_', '')) %>% 
  dplyr::select(GEO_accession:Slide) %>% 
  column_to_rownames(var='Sample_nameToRowname')

write_csv(GSE75196.meta_v1, path = "/home/qianhui/DNAme/Process_decidual/GEO_dataSets/GSE75196/GSE75196.meta_v1.csv")

##getwd()
# get RGset for GSE66210
# GSE75196.RGset <- getRGset('GSE75196')

```


## `GEO_num[5]` from GEO database. 
```{r ReadData5}
registerDoParallel(cores = 10)

# establish a proper samplesheet for 
GSE75248.meta_v1 <- GSE75248.meta %>% 
  mutate(GEO_accession=geo_accession) %>% 
  mutate(Sample= str_replace_all(description, 'normal infant placenta tissue', 'Placenta')) %>% 
  mutate(Sample_name = str_replace_all(supplementary_file, c("_(Red|Grn).idat.gz"="", ".+/"=""))) %>% 
  mutate(Sample_nameToRowname=Sample_name) %>% 
  mutate(Disease = str_replace_all(`birth weight group:ch1`, 'AGA', 'Uncomplicated')) %>% 
  inset('Gestation', value=rep('NA', nrow(GSE75248.meta))) %>% 
  inset('Trimester', value=rep('Term', nrow(GSE75248.meta))) %>% 
  inset('Additional_Info', value=rep('NA', nrow(GSE75248.meta))) %>% 
  inset('Additional_Info_2', value=rep('NA', nrow(GSE75248.meta))) %>% 
  mutate(Fetal_Sex= `gender:ch1`) %>% 
  inset('ArrayType', value=rep('450K', nrow(GSE75248.meta))) %>% 
  inset('Study', value=rep('GSE75248', nrow(GSE75248.meta))) %>% 
  mutate(Array= str_extract(Sample_name, pattern = "R..C..")) %>% 
  mutate(Slide=  str_extract(Sample_name, pattern= "_(.*?)_")) %>% 
  mutate(Slide=  str_replace_all(Slide, '_', '')) %>% 
  dplyr::select(GEO_accession:Slide) %>% 
  column_to_rownames(var='Sample_nameToRowname')

write_csv(GSE75248.meta_v1, path = "/home/qianhui/DNAme/Process_decidual/GEO_dataSets/GSE75248/GSE75248.meta_v1.csv")

##getwd()
# get RGset for GSE66210
# GSE75248.RGset <- getRGset('GSE75248')

```

## `GEO_num[6]` from GEO database. GSE93429 has no IDAT files, so it was removed.

## `GEO_num[7]` GSE93208
```{r ReadData6}
registerDoParallel(cores = 10)

# establish a proper samplesheet for 
GSE93208.meta_v1 <- GSE93208.meta %>% 
  mutate(GEO_accession=geo_accession) %>% 
  mutate(Sample= str_replace_all(`tissue:ch1`, 'chorionic villus', 'Placenta')) %>% 
  mutate(Sample_name = str_replace_all(supplementary_file, c("_(Red|Grn).idat.gz"="", ".+/"=""))) %>% 
  mutate(Sample_nameToRowname=Sample_name) %>% 
  inset('Disease', value = rep('Uncomplicated', nrow(GSE93208.meta))) %>% 
  # mutate(Gestation= str_replace_all(`gestational age (weeks):ch1`, c('8 to 10'='9', '12 to 14'='13'))) %>%
  mutate(Gestation= `gestational age (weeks):ch1`) %>%
  inset('Trimester', value=ifelse(GSE93208.meta$`gestational age (weeks):ch1`=='8 to 10', 'First', 'Second')) %>% 
  mutate(Additional_Info = title) %>% 
  inset('Additional_Info_2', value=rep('NA', nrow(GSE93208.meta))) %>%  
  inset('Fetal_Sex', value=rep('NA', nrow(GSE93208.meta))) %>% 
  inset('ArrayType', value=rep('450K', nrow(GSE93208.meta))) %>% 
  inset('Study', value=rep('GSE93208', nrow(GSE93208.meta))) %>% 
  mutate(Array= str_extract(Sample_name, pattern = "R..C..")) %>% 
  mutate(Slide=  str_extract(Sample_name, pattern= "_(.*?)_")) %>% 
  mutate(Slide=  str_replace_all(Slide, '_', '')) %>% 
  dplyr::select(GEO_accession:Slide) %>% 
  column_to_rownames(var='Sample_nameToRowname')

write_csv(GSE93208.meta_v1, path = "/home/qianhui/DNAme/Process_decidual/GEO_dataSets/GSE93208/GSE93208.meta_v1.csv")

##getwd()
# get RGset for GSE66210
# GSE93208.RGset <- getRGset('GSE93208')

```


## `GEO_num[8]` GSE115508
additional info is based on pathology groups.
```{r ReadData7}
registerDoParallel(cores = 10)

# establish a proper samplesheet for 
GSE115508.meta_v1 <- GSE115508.meta %>% 
  mutate(GEO_accession=geo_accession) %>% 
  mutate(Sample= str_replace_all(`tissue:ch1`, 'Chorionic villi', 'Placenta')) %>% 
  mutate(Sample_name = str_replace_all(supplementary_file, c("_(Red|Grn).idat.gz"="", ".+/"=""))) %>% 
  mutate(Sample_nameToRowname=Sample_name) %>% 
  mutate(Disease = str_replace_all(`pathology group:ch1`, 'non_chorioamnionitis', 'Uncomplicated')) %>% 
  mutate(Gestation= `gestational age:ch1`) %>% 
  inset('Trimester', value=ifelse(GSE115508.meta$`gestational age:ch1` >= 37, 'Term', 'Third')) %>% 
  inset('Additional_Info', value=rep('NA', nrow(GSE115508.meta))) %>%  
  inset('Additional_Info_2', value=rep('NA', nrow(GSE115508.meta))) %>%  
  mutate(Fetal_Sex= str_replace_all(`fetal sex:ch1`, c('F'='Female', 'M'='Male'))) %>% 
  inset('ArrayType', value=rep('EPIC', nrow(GSE115508.meta))) %>% 
  inset('Study', value=rep('GSE115508', nrow(GSE115508.meta))) %>% 
  mutate(Array= str_extract(Sample_name, pattern = "R..C..")) %>% 
  mutate(Slide=  str_extract(Sample_name, pattern= "_(.*?)_")) %>% 
  mutate(Slide=  str_replace_all(Slide, '_', '')) %>% 
  dplyr::select(GEO_accession:Slide) %>% 
  column_to_rownames(var='Sample_nameToRowname')

write_csv(GSE115508.meta_v1, path = "/home/qianhui/DNAme/Process_decidual/GEO_dataSets/GSE115508/GSE115508.meta_v1.csv")

##getwd()
# get RGset for GSE66210
# GSE115508.RGset <- getRGset('GSE115508')

```


## `GEO_num[9]` GSE120250

outlier information provided in Additional_Info column
```{r ReadData8}
registerDoParallel(cores = 10)

# establish a proper samplesheet for 
GSE120250.meta_v1 <- GSE120250.meta %>% 
  mutate(GEO_accession=geo_accession) %>% 
  mutate(Sample= str_replace_all(`title`, c('ART_..+'='Placenta', 'Cont..+'='Placenta'))) %>% 
  mutate(Sample_name = str_replace_all(supplementary_file, c("_(Red|Grn).idat.gz"="", ".+/"=""))) %>% 
  mutate(Sample_nameToRowname=Sample_name) %>% 
  mutate(Disease = str_replace_all(`characteristics_ch1.2`, 'art treatment: NA', 'Uncomplicated')) %>% 
  inset('Gestation', value=rep('NA', nrow(GSE120250.meta))) %>% 
  inset('Trimester', value=rep('Term', nrow(GSE120250.meta))) %>% 
  mutate(Additional_Info= str_replace_all(`outlier:ch1`, c('N'='Standard', 'Y'='Outlier'))) %>%  #outliers
  inset('Additional_Info_2', value=rep('NA', nrow(GSE120250.meta))) %>%  
  mutate(Fetal_Sex= str_replace_all(`gender:ch1`, c('F'='Female', 'M'='Male'))) %>% 
  inset('ArrayType', value=rep('450K', nrow(GSE120250.meta))) %>% 
  inset('Study', value=rep('GSE120250', nrow(GSE120250.meta))) %>% 
  mutate(Array= str_extract(Sample_name, pattern = "R..C..")) %>% 
  mutate(Slide=  str_extract(Sample_name, pattern= "_(.*?)_")) %>% 
  mutate(Slide=  str_replace_all(Slide, '_', '')) %>% 
  dplyr::select(GEO_accession:Slide) %>% 
  column_to_rownames(var='Sample_nameToRowname')

write_csv(GSE120250.meta_v1, path = "/home/qianhui/DNAme/Process_decidual/GEO_dataSets/GSE120250/GSE120250.meta_v1.csv")

##getwd()
# get RGset for GSE66210
# GSE120250.RGset <- getRGset('GSE120250')

```


## `GEO_num[10]` GSE108567 (samples from published data sets, could be redundant samples)
Additional_Info: relation to previous GSM sample ID
additional_info_2: revised processing group
```{r ReadData9}
# registerDoParallel(cores = 10)
# 
# # establish a proper samplesheet for 
# GSE108567.meta_v1 <- GSE108567.meta %>% 
#   mutate(GEO_accession=geo_accession) %>% 
#   mutate(Sample= `tissue:ch1`) %>% 
#   mutate(Sample_name = str_replace_all(supplementary_file, c("_(Red|Grn).idat.gz"="", ".+/"=""))) %>% 
#   mutate(Sample_nameToRowname=Sample_name) %>% 
#   inset('Disease', value = rep('NA', nrow(GSE108567.meta))) %>% 
#   mutate(Gestation= `gestational age (week, rounded):ch1`) %>% 
#   inset('Trimester', value=ifelse(GSE108567.meta$`gestational age (week, rounded):ch1` >= 37, 'Term', 'Third')) %>% 
#   mutate(Additional_Info = `relation`) %>% 
#   mutate(Additional_Info_2= `processing group:ch1`) %>% 
#   mutate(Fetal_Sex= `fetal sex:ch1`) %>% 
#   inset('ArrayType', value=rep('450K', nrow(GSE108567.meta))) %>% 
#   inset('Study', value=rep('GSE108567', nrow(GSE108567.meta))) %>% 
#   mutate(Array= str_extract(Sample_name, pattern = "R..C..")) %>% 
#   mutate(Slide=  str_extract(Sample_name, pattern= "_(.*?)_")) %>% 
#   mutate(Slide=  str_replace_all(Slide, '_', '')) %>% 
#   dplyr::select(GEO_accession:Slide) %>% 
#   column_to_rownames(var='Sample_nameToRowname')
# 
# write_csv(GSE108567.meta_v1, path = "/home/qianhui/DNAme/Process_decidual/GEO_dataSets/GSE108567/GSE108567.meta_v1.csv")
# 
# ##getwd()
# # get RGset for GSE66210
# # GSE108567.RGset <- getRGset('GSE108567')

```

## `GEO_num[11]` GSE100197 
Additional_Info: pathology group
additional_info_2: sample tissue
```{r ReadData10}
registerDoParallel(cores = 10)

# establish a proper samplesheet for 
GSE100197.meta_v1 <- GSE100197.meta %>% 
  mutate(GEO_accession=geo_accession) %>% 
  mutate(Sample= str_replace_all(`sample tissue:ch1`, 'Chorionic..+', 'Placenta')) %>%  
  mutate(Sample_name = str_replace_all(supplementary_file, c("_(Red|Grn).idat.gz"="", ".+/"=""))) %>% 
  mutate(Sample_nameToRowname=Sample_name) %>% 
  mutate(Disease = str_replace_all(`pathology group:ch1`, c('REPLICATE'='TechnicalReplicate', 'Term'='Uncomplicated'))) %>% 
  mutate(Gestation= `gestational age:ch1`) %>% 
  inset('Trimester', value=ifelse(GSE100197.meta$`gestational age:ch1` >= 37, 'Term', 'Third')) %>% 
  mutate(Additional_Info = `sample tissue:ch1`) %>% 
  inset('Additional_Info_2', value=rep('NA', nrow(GSE100197.meta))) %>%  
  mutate(Fetal_Sex= str_replace_all(`fetal sex:ch1`, c('FEMALE'='Female', 'MALE'='Male'))) %>% 
  inset('ArrayType', value=rep('450K', nrow(GSE100197.meta))) %>% 
  inset('Study', value=rep('GSE100197', nrow(GSE100197.meta))) %>% 
  mutate(Array= str_extract(Sample_name, pattern = "R..C..")) %>% 
  mutate(Slide=  str_extract(Sample_name, pattern= "[^_]+_[^_]+")) %>% 
  # mutate(Slide=  str_replace_all(Slide, '_', '')) %>% 
  dplyr::select(GEO_accession:Slide) %>% 
  column_to_rownames(var='Sample_nameToRowname')

write_csv(GSE100197.meta_v1, path = "/home/qianhui/DNAme/Process_decidual/GEO_dataSets/GSE100197/GSE100197.meta_v1.csv")

##getwd()
# get RGset for GSE66210
# GSE100197.RGset <- getRGset('GSE100197')

```


## `GEO_num[12]` GSE98224, 48 samples are methylation samples, other 48 are expression arrary data 
Additional_Info: previous hypertensive pregnancy
additional_info_2: previous miscarriage
```{r ReadData11}
registerDoParallel(cores = 10)

# establish a proper samplesheet for 
GSE98224.meta_v1 <- GSE98224.meta %>% 
  mutate(GEO_accession=geo_accession) %>% 
  mutate(Sample= str_replace_all(`tissue:ch1`, 'placenta', 'Placenta')) %>% 
  mutate(Sample_name = str_replace_all(supplementary_file, c("_(Red|Grn).idat.gz"="", ".+/"=""))) %>% 
  mutate(Sample_nameToRowname=Sample_name) %>%
  mutate(Disease = str_replace_all(`diagnosis:ch1`, 'non-PE','Uncomplicated')) %>% 
  mutate(Gestation= `ga week:ch1`) %>% 
  inset('Trimester', value=ifelse(GSE98224.meta$`ga week:ch1` >= 37, 'Term', 'Third')) %>% 
  mutate(Additional_Info = `previous hypertensive pregnancy:ch1`) %>% 
  mutate(Additional_Info_2= `previous miscarriage:ch1`) %>%  
  mutate(Fetal_Sex= str_replace_all(`gender:ch1`, c('F'='Female', 'M'='Male'))) %>% 
  inset('ArrayType', value=rep('450K', nrow(GSE98224.meta))) %>% 
  inset('Study', value=rep('GSE98224', nrow(GSE98224.meta))) %>% 
  mutate(Array= str_extract(Sample_name, pattern = "R..C..")) %>% 
  mutate(Slide=  str_extract(Sample_name, pattern= "_(.*?)_")) %>% 
  mutate(Slide=  str_replace_all(Slide, '_', '')) %>% 
  dplyr::select(GEO_accession:Slide) %>% 
  column_to_rownames(var='Sample_nameToRowname')

write_csv(GSE98224.meta_v1, path = "/home/qianhui/DNAme/Process_decidual/GEO_dataSets/GSE98224/GSE98224.meta_v1.csv")

##getwd()
# get RGset for GSE66210
# GSE98224.RGset <- getRGset('GSE98224')

```


## 13. GSE106089 `GEO_num[12]`
Additional_Info: antibiotic use
additional_info_2: NA
```{r ReadData12}
registerDoParallel(cores = 10)

# establish a proper samplesheet for 
GSE106089.meta_v1 <- GSE106089.meta %>% 
  mutate(GEO_accession=geo_accession) %>% 
  mutate(Sample= str_replace_all(description, 'normal placental sample', 'Placenta')) %>% 
  mutate(Sample_name = str_replace_all(supplementary_file, c("_(Red|Grn).idat.gz"="", ".+/"=""))) %>% 
  mutate(Sample_nameToRowname=Sample_name) %>% 
  inset('Disease', value = ifelse(GSE106089.meta$`bacteria present:ch1`=='None', 'Uncomplicated', 'BacteriaDetected')) %>% 
  inset('Gestation', value=rep('NA', nrow(GSE106089.meta))) %>% 
  inset('Trimester', value=rep('Third', nrow(GSE106089.meta))) %>% 
  mutate(Additional_Info = `characteristics_ch1.4`) %>% 
  inset('Additional_Info_2', value=rep('NA', nrow(GSE106089.meta))) %>%  
  mutate(Fetal_Sex= str_replace_all(`fetal sex:ch1`, c('^male'='Male','^female'='Female'))) %>% 
  inset('ArrayType', value=rep('450K', nrow(GSE106089.meta))) %>% 
  inset('Study', value=rep('GSE106089', nrow(GSE106089.meta))) %>% 
  mutate(Array= str_extract(Sample_name, pattern = "R..C..")) %>% 
  mutate(Slide=  str_extract(Sample_name, pattern= "_(.*?)_")) %>% 
  mutate(Slide=  str_replace_all(Slide, '_', '')) %>% 
  dplyr::select(GEO_accession:Slide) %>% 
  column_to_rownames(var='Sample_nameToRowname')

write_csv(GSE106089.meta_v1, path = "/home/qianhui/DNAme/Process_decidual/GEO_dataSets/GSE106089/GSE106089.meta_v1.csv")

##getwd()
# get RGset for GSE66210
# GSE106089.RGset <- getRGset('GSE106089')

```


## 14. GSE71678, `GEO_num[13]` the meta data is wrong in `hyb_protocol` coplumn, it should be 450k not 27k.
Additional_Info: NA
additional_info_2: NA
```{r ReadData13}
registerDoParallel(cores = 10)

# establish a proper samplesheet for 
GSE71678.meta_v1 <- GSE71678.meta %>% 
  mutate(GEO_accession=geo_accession) %>% 
  mutate(Sample= str_replace_all(description, 'normal infant placenta tissue', 'Placenta')) %>% 
  mutate(Sample_name = str_replace_all(supplementary_file, c("_(Red|Grn).idat.gz"="", ".+/"=""))) %>% 
  mutate(Sample_nameToRowname=Sample_name) %>% 
  inset('Disease', value = ifelse(GSE71678.meta$`placental as levels:ch1`=='NA', 'Uncomplicated', 'ArsenicDetected')) %>% 
  mutate(Gestation = `gestational age:ch1`) %>% 
  inset('Trimester', value=ifelse(GSE71678.meta$`gestational age:ch1` >= 37, 'Term', 'Third')) %>% 
  inset('Additional_Info', value=rep('NA', nrow(GSE71678.meta))) %>%  
  inset('Additional_Info_2', value=rep('NA', nrow(GSE71678.meta))) %>%  
  mutate(Fetal_Sex= `infant gender:ch1`) %>% 
  inset('ArrayType', value=rep('450K', nrow(GSE71678.meta))) %>% 
  inset('Study', value=rep('GSE71678', nrow(GSE71678.meta))) %>% 
  mutate(Array= str_extract(Sample_name, pattern = "R..C..")) %>% 
  mutate(Slide=  str_extract(Sample_name, pattern= "_(.*?)_")) %>% 
  mutate(Slide=  str_replace_all(Slide, '_', '')) %>% 
  dplyr::select(GEO_accession:Slide) %>% 
  column_to_rownames(var='Sample_nameToRowname')

write_csv(GSE71678.meta_v1, path = "/home/qianhui/DNAme/Process_decidual/GEO_dataSets/GSE71678/GSE71678.meta_v1.csv")

##getwd()
# get RGset for GSE66210
# GSE71678.RGset <- getRGset('GSE71678')

```


## 15. GSE79262 
There is no raw IDAT files for this data set, so we can't use it.
```{r ReadData, eval=FALSE}
registerDoParallel(cores = 10)
# establish a proper samplesheet for 
IDATfileNames <- list.files(path = "~/DNAme/Process_decidual/GEO_dataSets/GSE79262", pattern = ".idat.gz")
sampleNames <- str_replace_all(IDATfileNames, "_(Red|Grn).idat.gz", "") %>% unique()
# matchedSampleNames <- sampleNames[match(GSE79262.meta$geo_accession, str_extract(sampleNames, pattern = '[^_]+'))]

```


## 16. GSE113600 `GEO_num[14]`
Additional_Info: NA
additional_info_2: NA
```{r ReadData14}
registerDoParallel(cores = 10)

# establish a proper samplesheet for 
GSE113600.meta_v1 <- GSE113600.meta %>% 
  mutate(GEO_accession=geo_accession) %>% 
  mutate(Sample= str_replace_all(`tissue:ch1`, 'decidua', 'Decidua')) %>% 
  mutate(Sample_name = str_replace_all(supplementary_file, c("_(Red|Grn).idat.gz"="", ".+/"=""))) %>% 
  mutate(Sample_nameToRowname=Sample_name) %>% 
  mutate(Disease = str_replace_all(`condition:ch1`, 'induced abortion', 'Uncomplicated')) %>% 
  inset('Gestation', value=rep('NA', nrow(GSE113600.meta))) %>% 
  inset('Trimester', value=rep('First', nrow(GSE113600.meta))) %>% 
  inset('Additional_Info', value=rep('NA', nrow(GSE113600.meta))) %>%  
  inset('Additional_Info_2', value=rep('NA', nrow(GSE113600.meta))) %>%  
  inset('Fetal_Sex', value=rep('NA', nrow(GSE113600.meta))) %>% 
  inset('ArrayType', value=rep('EPIC', nrow(GSE113600.meta))) %>% 
  inset('Study', value=rep('GSE113600', nrow(GSE113600.meta))) %>% 
  mutate(Array= str_extract(Sample_name, pattern = "decidua.")) %>% 
  mutate(Slide= str_extract(Sample_name, pattern = "[^_]+")) %>%
  dplyr::select(GEO_accession:Slide) %>% 
  column_to_rownames(var='Sample_nameToRowname')

write_csv(GSE113600.meta_v1, path = "/home/qianhui/DNAme/Process_decidual/GEO_dataSets/GSE113600/GSE113600.meta_v1.csv")

# GSE113600.RGset <- getRGset('GSE113600')
# phenotypes <- read.metharray.sheet('/home/qianhui/DNAme/Process_decidual/GEO_dataSets/GSE113600', pattern = "_v1.csv$")


```


## 17. GSE41933, `GEO_num[15]`, there's no raw IDAT files for this data set, so we didn't use it

## 18. GSE66459, `GEO_num[16]`
Additional_Info: NA
additional_info_2: NA
```{r ReadData15}
registerDoParallel(cores = 10)

# establish a proper samplesheet for 
GSE66459.meta_v1 <- GSE66459.meta %>% 
  mutate(GEO_accession=geo_accession) %>% 
  mutate(Sample= `tissue:ch1`) %>% 
  mutate(Sample_name = str_replace_all(supplementary_file, c("_(Red|Grn).idat.gz"="", ".+/"=""))) %>% 
  mutate(Sample_nameToRowname=Sample_name) %>% 
  inset('Disease', value = rep('Uncomplicated', nrow(GSE66459.meta))) %>% 
  mutate(Gestation = as.numeric(`gestational_age (days):ch1`)/7) %>% 
  inset('Trimester', value=ifelse(.$Gestation >= 37, 'Term', 'Third')) %>% 
  inset('Additional_Info', value=rep('NA', nrow(GSE66459.meta))) %>%  
  inset('Additional_Info_2', value=rep('NA', nrow(GSE66459.meta))) %>%  
  mutate(Fetal_Sex = `gender:ch1`) %>% 
  inset('ArrayType', value=rep('450K', nrow(GSE66459.meta))) %>% 
  inset('Study', value=rep('GSE66459', nrow(GSE66459.meta))) %>% 
  mutate(Array= str_extract(Sample_name, pattern = '(_.|_..)$')) %>% 
  mutate(Array= str_replace_all(Array, '_', '')) %>% 
  mutate(Slide=  str_extract(Sample_name, pattern= "_(.*?)_")) %>% 
  mutate(Slide=  str_replace_all(Slide, '_', '')) %>% 
  dplyr::select(GEO_accession:Slide) %>% 
  column_to_rownames(var='Sample_nameToRowname')

write_csv(GSE66459.meta_v1, path = "/home/qianhui/DNAme/Process_decidual/GEO_dataSets/GSE66459/GSE66459.meta_v1.csv")

# GSE66459.RGset <- getRGset('GSE66459')

```

## 19. our data
Add our outliers 5 + 5 normal placenta
```{r ReadData16}

baseDirOurEPIC <- file.path("/home/qianhui/DNAme/EPICarrayData/idat_files")
phenotypes <- read.metharray.sheet(baseDirOurEPIC)

set.seed(2018)
phenotypes_GA6to9standard <- phenotypes %>% 
  dplyr::filter(Pool_ID%in%"Placental Tissue") %>% 
  dplyr::filter(Sample_Group%in%c(6:9)) %>% 
  dplyr::filter(Outliers%in%"Standard") %>% 
  sample_n(5)

phenotypes_GA6to9_10samples_inGEOformat <-  phenotypes %>% 
  dplyr::filter(Pool_ID%in%"Placental Tissue") %>% 
  dplyr::filter(Sample_Group%in%c(6:9)) %>% 
  dplyr::filter(Outliers%in%"Outlier") %>% 
  bind_rows(phenotypes_GA6to9standard) %>% 
  inset('GEO_accession', value=rep("OurData", 10)) %>% 
  inset('Sample', value=rep('Placenta', 10)) %>% 
  mutate(Sample_name = str_replace_all(Basename, ".+/", "")) %>% 
  inset('Disease', value=rep("Uncomplicated", 10)) %>% 
  mutate(Gestation= as.character(Gestational.Age)) %>% 
  mutate(Trimester= str_replace_all(Trimester, "1st-Trimester", "First")) %>% 
  mutate(Additional_Info =Outliers) %>% 
  inset('Additional_Info_2', value=rep("NA", 10)) %>% 
  mutate(Fetal_Sex=str_replace_all(Fetal.Sex, c('M'='Male', 'F'='Female'))) %>% 
  inset('ArrayType', value=rep('EPIC',10)) %>% 
  inset('Study', value=rep('ourStudy', 10)) %>% 
  dplyr::select(GEO_accession:Gestation, Trimester, Additional_Info:Study, Array, Slide)

# I want to copy these 10 samples to IDATall folder, so I need to write this txt first 
write.table(phenotypes_GA6to9_10samples_inGEOformat$Sample_name, 
            file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/ourStudy10samples.txt", 
            quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

# bind our phenodata with GEO meta data

save(GSE66210.meta_v1, GSE74738.meta_v1, GSE69502.meta_v1, GSE75196.meta_v1, 
     GSE75248.meta_v1, GSE93208.meta_v1, GSE115508.meta_v1, GSE120250.meta_v1, 
     GSE100197.meta_v1, GSE98224.meta_v1, GSE106089.meta_v1, GSE71678.meta_v1, 
     GSE113600.meta_v1, GSE66459.meta_v1, phenotypes_GA6to9_10samples_inGEOformat,
     file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_allMetaData_v1.RData")

```


# 6. bind all the meta data together and read `GEOmeta` samplesheet. 
  Then read the data we wanted to Red-Green channel sets.
```{r GEO_meta_all}

load(file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_allMetaData_v1.RData")

GEOmeta <- rbind(GSE66210.meta_v1, GSE74738.meta_v1, GSE69502.meta_v1, GSE75196.meta_v1, 
                 GSE75248.meta_v1, GSE93208.meta_v1, GSE115508.meta_v1, GSE120250.meta_v1, 
                 GSE100197.meta_v1, GSE98224.meta_v1, GSE106089.meta_v1, GSE71678.meta_v1, 
                 GSE113600.meta_v1, GSE66459.meta_v1, phenotypes_GA6to9_10samples_inGEOformat)

# remove the samples with no trimester informations
GEOmeta <- GEOmeta[!(GEOmeta$Trimester%in%"NA"), ]

write_csv(GEOmeta, path = "/home/qianhui/DNAme/Process_decidual/GEO_dataSets/IDAT_all/GEOmeta.csv")

GEO_baseDir <- file.path("/home/qianhui/DNAme/Process_decidual/GEO_dataSets/IDAT_all")

GEO_phenotypes <- read.metharray.sheet(GEO_baseDir, pattern = "GEOmeta.csv$")

save(GEOmeta, GEO_phenotypes, 
     file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_phenotypes_add10.RData")

```


## read uncomplicated placental sampels 

We read 450K and EPIC data separately, because they have different annotations.

### load phenodata, start from here, so we don't need to redo all steps above.
```{r}

load(file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_phenotypes_add10.RData")

# filter out some data sets /samples that we don't need
## remove complicated samples by Disease column (-864)
GEO_phenotypes_uncom <- GEO_phenotypes[GEO_phenotypes$Disease%in%"Uncomplicated",]

## remove samples that were replicate samples in GSE66210 (-4)
Du <- c("GSM1617040","GSM1617041", "GSM1617042","GSM1617043")
GEO_phenotypes_mGSE66210 <- GEO_phenotypes_uncom[!(GEO_phenotypes_uncom$Study%in%"GSE66210" & GEO_phenotypes_uncom$GEO_accession%in%Du),]

## remove perterm samples from GSE74738 (-15)
GEO_phenotypes_mGSE74738 <- GEO_phenotypes_mGSE66210 %>% 
  dplyr::filter(!(Study%in%"GSE74738" & Sample%in%c("placental mesenchyme", "placental trophoblast"))) %>% 
  dplyr::filter(!(Study%in%"GSE74738" & Trimester%in%"Third"))
  
## remove samples from GSE93208 (these samples were isolated cells not bulk tissue) (-19)
## and GSE106089 (these samples were all preterm) (-46), so -65 intotal.
GEO_phenotypes_mGSE93208 <- GEO_phenotypes_mGSE74738 %>% 
  dplyr::filter(!(GEO_phenotypes_mGSE74738$Study%in%c("GSE93208","GSE106089")))

## remove preterm samples from GSE98224 (-9)
GEO_phenotypes_mGSE98224 <- GEO_phenotypes_mGSE93208 %>% 
  dplyr::filter(!(Study%in%"GSE98224" & Trimester%in%"Third"))

## remove preterm samples from GSE71678 (-4)
GEO_phenotypes_mGSE71678 <- GEO_phenotypes_mGSE98224 %>% 
  dplyr::filter(!(Study%in%"GSE71678" & Trimester%in%"Third"))

## remove preterm samples from GSE66459 (-11)
GEO_phenotypes_mGSE66459 <- GEO_phenotypes_mGSE71678 %>% 
  dplyr::filter(!(Study%in%"GSE66459" & Trimester%in%"Third"))

## remove preterm samples from GSE115508 (-35)
GEO_phenotypes_mGSE115508 <- GEO_phenotypes_mGSE66459 %>% 
  dplyr::filter(!(Study%in%"GSE115508" & Trimester%in%"Third"))

## remove samples that were not placenta in GSE69502 (-49)
GEO_phenotypes_mGSE69502 <- GEO_phenotypes_mGSE115508 %>% 
  dplyr::filter(!(Study%in%"GSE69502" & !(Sample%in%"Placenta")))

GEO_phenotypes_Filtered <- GEO_phenotypes_mGSE69502

# saveRDS(GEO_phenotypes_Filtered, 
#      file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_phenotypes410_Filtered.rds")

```

### read array datas
```{r normalPlacenta}
  
# read Red-Green set for 450K samples
GEO_phenotypes_450k <- GEO_phenotypes_Filtered[GEO_phenotypes_Filtered$ArrayType=='450K',]

# GEO_phenotypes_450k_duplicatedBasename <- GEO_phenotypes_450k[duplicated(GEO_phenotypes_450k$Basename),]
# GEO_phenotypes_450k_allDuplicates <- GEO_phenotypes_450k[GEO_phenotypes_450k$Basename%in%GEO_phenotypes_450k_duplicatedBasename$Basename,]
# GEO_phenotypes_450k_dupGSE74738 <- GEO_phenotypes_450k_allDuplicates %>% filter(Study=='GSE74738')
# GEO_phenotypes_450k_dupGSE74738[match(GEO_phenotypes_450k_duplicatedBasename$Slide, GEO_phenotypes_450k_dupGSE74738$Slide),]
# GEO_phenotypes_450k_removeDup <- GEO_phenotypes_450k[!(GEO_phenotypes_450k$GEO_accession %in%
#                                                      GEO_phenotypes_450k_duplicatedBasename$GEO_accession),]

GEO_phenotypes_450k_normalPlacenta <- GEO_phenotypes_450k %>% 
  filter(str_detect(Sample, pattern="^Placenta$")) %>% 
  filter(str_detect(Disease, pattern="Uncomplicated$"))

# there are `nrow(GEO_phenotypes_450k_normalPlacenta)` normal placenta samples in total
GEO_RGset_450k_normalPlacenta <- read.metharray.exp(base = GEO_baseDir, targets = GEO_phenotypes_450k_normalPlacenta)
GEO_phenoData_450k_normalPlacenta <- pData(GEO_RGset_450k_normalPlacenta) #chain data & phenotypes together


# read RG set fot EPIC samples
## read our study
GEO_phenotypes_EPIC_ourStudy <- GEO_phenotypes_Filtered[GEO_phenotypes_Filtered$ArrayType%in%'EPIC' & GEO_phenotypes_Filtered$Study%in%'ourStudy',]
## there are 10 normal placenta samples in total random selected from our study 
GEO_RGset_EPIC_normalPlacenta_ourStudy <- read.metharray.exp(base = GEO_baseDir, targets = GEO_phenotypes_EPIC_ourStudy)
GEO_phenoData_EPIC_normalPlacenta_ourStudy <- pData(GEO_RGset_EPIC_normalPlacenta_ourStudy)

## read other EPIC data, except our study
GEO_phenotypes_EPIC <- GEO_phenotypes_Filtered[GEO_phenotypes_Filtered$ArrayType%in%'EPIC' & !(GEO_phenotypes_Filtered$Study%in%'ourStudy'),]

GEO_phenotypes_EPIC_normalPlacenta <- GEO_phenotypes_EPIC %>% 
  filter(str_detect(Sample, pattern="^Placenta$")) %>% 
  filter(str_detect(Disease, pattern="Uncomplicated$"))

## there are 25 normal placenta samples in total, except our 10 samples
GEO_RGset_EPIC_normalPlacenta <- read.metharray.exp(base = GEO_baseDir, targets = GEO_phenotypes_EPIC_normalPlacenta)
GEO_phenoData_EPIC_normalPlacenta <- pData(GEO_RGset_EPIC_normalPlacenta) #chain data & phenotypes together

# save all outputs
save(GEO_RGset_450k_normalPlacenta, GEO_phenoData_450k_normalPlacenta, 
     GEO_RGset_EPIC_normalPlacenta, GEO_phenoData_EPIC_normalPlacenta, 
     GEO_RGset_EPIC_normalPlacenta_ourStudy, GEO_phenoData_EPIC_normalPlacenta_ourStudy,
     file ="/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_RG&pData_normalPlacenta.RData")

```



## read other tissue samples, also uncomplicated samples  

 * GSE74738, 450K data, could be gold standard for different tissue types.

### functions for getting RGset--the raw data set for each tissue types.

we write 2 functions, one for getting RGset form 450k data, another for getting RGset from EPIC data.

```{r function2ReadFiles}
GEOmeta$Sample %>% table()
GEO_phenotypes_450k$Sample %>% table()
GEO_phenotypes_EPIC$Sample %>% table()
GEO_phenotypes_EPIC_ourStudy$Sample %>% table()

# 2 functions to get RGset from 450K and EPIC platform:
## get normal tissue data, 450K
getRGset450K <- function(Tissue){
  # base directory
  GEO_baseDir <- file.path("/home/qianhui/DNAme/Process_decidual/GEO_dataSets/IDAT_all")
  # get 450k meta data
  GEO_phenotypes_450k <- GEO_phenotypes_Filtered[GEO_phenotypes_Filtered$ArrayType=='450K',]
  # get 450k, meta data for the tissue
  GEO_phenotypes_450k_Tissue <- GEO_phenotypes_450k %>% 
  filter(str_detect(Sample, pattern=paste0("^", Tissue, "$"))) %>% 
  filter(str_detect(Disease, pattern="Uncomplicated$"))
  # get the RGset for this tissue
  GEO_RGset_450k_Tissue <- read.metharray.exp(base = GEO_baseDir, targets = GEO_phenotypes_450k_Tissue)
  # GEO_phenoData_450k_Tissue <- pData(GEO_RGset_450k_Tissue) #chain data & phenotypes together
  
  return(GEO_RGset_450k_Tissue)
}

## get normal tissue data, EPIC
getRGsetEPIC <- function(Tissue){
  # base directory
  GEO_baseDir <- file.path("/home/qianhui/DNAme/Process_decidual/GEO_dataSets/IDAT_all")
  # get 450k meta data
  GEO_phenotypes_EPIC <- GEO_phenotypes_Filtered[GEO_phenotypes_Filtered$ArrayType=='EPIC',]
  # get 450k, meta data for the tissue
  GEO_phenotypes_EPIC_Tissue <- GEO_phenotypes_EPIC %>% 
  filter(str_detect(Sample, pattern=paste0("^", Tissue, "$"))) %>% 
  filter(str_detect(Disease, pattern="Uncomplicated$"))
  # get the RGset for this tissue
  GEO_RGset_EPIC_Tissue <- read.metharray.exp(base = GEO_baseDir, targets = GEO_phenotypes_EPIC_Tissue)
  # GEO_phenoData_450k_Tissue <- pData(GEO_RGset_450k_Tissue) #chain data & phenotypes together
  
  return(GEO_RGset_EPIC_Tissue)
}


```


### other 8 tissue types:
The other 8 tissue types except for placenta, including:
Amnoin, Chorion, Decidua, maternal whole blood, placental mesenchyme, placental trophoblast, umbilical cord and umbilical cord blood.

#### 1) Amnion
```{r Amnion}
# Amnion
## 450k
GEO_RGset_450k_Amnion <- getRGset450K(Tissue = "Amnion")
GEO_phenoData_450k_Amnion <- pData(GEO_RGset_450k_Amnion) #chain data & phenotypes together
## EPIC
# GEO_RGset_EPIC_Amnion <- getRGsetEPIC(Tissue = "Amnion")
# GEO_phenoData_EPIC_Amnion <- pData(GEO_RGset_EPIC_Amnion) #chain data & phenotypes together
## save outputs for this tissue
save(GEO_RGset_450k_Amnion, GEO_phenoData_450k_Amnion, 
     # GEO_RGset_EPIC_Amnion, GEO_phenoData_EPIC_Amnion, 
     file ="/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_RG&pData_normalAmnion.RData")

```

#### 2) Chorion
```{r Chorion}
## 450k
GEO_RGset_450k_Chorion <- getRGset450K(Tissue = "Chorion")
GEO_phenoData_450k_Chorion <- pData(GEO_RGset_450k_Chorion) #chain data & phenotypes together
## EPIC
# GEO_RGset_EPIC_Chorion <- getRGsetEPIC(Tissue = "Chorion")
# GEO_phenoData_EPIC_Chorion <- pData(GEO_RGset_EPIC_Chorion) #chain data & phenotypes together
## save outputs for this tissue
save(GEO_RGset_450k_Chorion, GEO_phenoData_450k_Chorion, 
     # GEO_RGset_EPIC_Chorion, GEO_phenoData_EPIC_Chorion, 
     file ="/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_RG&pData_normalChorion.RData")

```


#### 3) Decidua
```{r Decidua}
## 450k
GEO_RGset_450k_Decidua <- getRGset450K(Tissue = "Decidua")
GEO_phenoData_450k_Decidua <- pData(GEO_RGset_450k_Decidua) #chain data & phenotypes together

## EPIC
GEO_RGset_EPIC_Decidua <- getRGsetEPIC(Tissue = "Decidua")
GEO_phenoData_EPIC_Decidua <- pData(GEO_RGset_EPIC_Decidua) #chain data & phenotypes together
## save outputs for this tissue
save(GEO_RGset_450k_Decidua, GEO_phenoData_450k_Decidua, 
     GEO_RGset_EPIC_Decidua, GEO_phenoData_EPIC_Decidua, 
     file ="/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_RG&pData_normalDecidua.RData")

```

#### 4) Maternal whole blood
```{r MaternalBlood}
## 450k
GEO_RGset_450k_MaternalBlood <- getRGset450K(Tissue = "Maternal whole blood")
GEO_phenoData_450k_MaternalBlood <- pData(GEO_RGset_450k_MaternalBlood) #chain data & phenotypes together

## save outputs for this tissue
save(GEO_RGset_450k_MaternalBlood, GEO_phenoData_450k_MaternalBlood, 
     # GEO_RGset_EPIC_MaternalBlood, GEO_phenoData_EPIC_MaternalBlood, 
     file ="/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_RG&pData_normalMaternalBlood.RData")

```


<!-- #### 5) Placental mesenchyme -->

<!-- ```{r PlacentalMesenchyme} -->
<!-- ## 450k -->
<!-- GEO_RGset_450k_mesenchyme <- getRGset450K(Tissue = "placental mesenchyme") -->
<!-- GEO_phenoData_450k_mesenchyme <- pData(GEO_RGset_450k_mesenchyme) #chain data & phenotypes together -->

<!-- ## save outputs for this tissue -->
<!-- save(GEO_RGset_450k_mesenchyme, GEO_phenoData_450k_mesenchyme,  -->
<!--      file ="/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_RG&pData_normalPlacentalMesenchyme.RData") -->

<!-- ``` -->


<!-- #### 6) Placental trophoblast -->

<!-- ```{r Trophoblast} -->
<!-- ## 450k -->
<!-- GEO_RGset_450k_trophoblast <- getRGset450K(Tissue = "placental trophoblast") -->
<!-- GEO_phenoData_450k_trophoblast <- pData(GEO_RGset_450k_trophoblast) #chain data & phenotypes together -->

<!-- ## save outputs for this tissue -->
<!-- save(GEO_RGset_450k_trophoblast, GEO_phenoData_450k_trophoblast,  -->
<!--      file ="/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_RG&pData_normalPlacentalTrophoblast.RData") -->

<!-- ``` -->


#<!-- #### 7) Umbilical cord -->

<!-- ```{r UmbilicalCord} -->

<!-- ## 450k -->
<!-- GEO_RGset_450k_UC <- getRGset450K(Tissue = "umbilical cord") -->
<!-- GEO_phenoData_450k_UC <- pData(GEO_RGset_450k_UC) #chain data & phenotypes together -->

<!-- ## save outputs for this tissue -->
<!-- save(GEO_RGset_450k_UC, GEO_phenoData_450k_UC,  -->
<!--      file ="/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_RG&pData_normalUmbilicalCord.RData") -->

<!-- ``` -->

#### 85 Umbilical cord blood

```{r CordBlood}

## 450k
GEO_RGset_450k_UCB <- getRGset450K(Tissue = "Umbilical cord blood")
GEO_phenoData_450k_UCB <- pData(GEO_RGset_450k_UCB) #chain data & phenotypes together

## save outputs for this tissue
save(GEO_RGset_450k_UCB, GEO_phenoData_450k_UCB, 
     file ="/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_RG&pData_normalUmbilicalCordBlood.RData")

```

