---
title: "GEO data sets_ICR analysis_sub"
author: "Qianhui"
date: "14/01/2019"
output: html_document
---

This document is mainly for part8 of the GEO data set analysis, that is match beta values with imprinting control regions (ICRs).


```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.align = "center", error = TRUE, echo = TRUE) # fig.width = 7, 
```


# 1. load data and packages
```{r package, warning=FALSE, message=FALSE}

library(tidyverse)
library(minfi)
library(magrittr)
library(tibble)
library(readxl)
library(doParallel)
library(Gviz)
library(dplyr)
library(reshape2)
library(minfi)
library(plyranges)
library(BSgenome.Hsapiens.UCSC.hg19)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(ggplot2)
library(ggpubr)
library(RColorBrewer)
pal <- brewer.pal(8, "Dark2")

# GEOmeta, GEO_phenotypes
load(file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/GEO_phenotypes_add10.RData")

# annotation needed for using functions
ann450k <- readRDS(file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/ann450k_GEO15sets.rds")
annEPIC <- readRDS(file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/annEPIC_GEO15sets.rds")

NormBMlist <- readRDS(file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/NormBMlist_300k_sub.rds")

# for annotating outliers
placentaOutlier <- readRDS(file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/placenta_beta_dim1Top2perc.df_all_sub.rds") %>%
  dplyr::select(Sample_name, Outlier) %>% 
  unique() %>% 
  `colnames<-`(c("Sample_name","Outlier3"))

```


# Import ICRs

There are 59 ICRs in GRange object `ICRs_v1`.
```{r}

ICRs <- read_xlsx(path = "/home/qianhui/DNAme/Process_decidual/ICRs/ICR_DMR_excelFile_fromPervjakova_etal.xlsx")

# other ICRs are regions without probes in those regions
ICR_Type <- c(rep('Ubiquitously imprinted', 45), rep('PlacentaSpecific', 14), rep('otherICRs', 8))

ICRsGR <- ICRs %>% na.omit() %>% inset('ICR_Type', value=ICR_Type) %>% 
           mutate(Gene=`Gene locus`) %>% 
           mutate(seqnames=paste0('chr', .$Chr)) %>% 
           mutate(start=Start) %>% 
           mutate(end=Finish) %>% 
           mutate(MethylationOrigin=`Methylation origin`) %>% 
           mutate(MeanBetaValue=str_replace_all(`Infinium beta value (min-max)`, " [(].+[)]", "")) %>% 
           dplyr::select(ICR_Type:MeanBetaValue) %>% 
           dplyr::filter(ICR_Type!="otherICRs") %>% 
           mutate(start = replace(start, which(Gene == "NESP-AS/GNAS-AS1"), 57425649)) %>% 
           as_granges()

# match ICRs with the genes with 
genes <- annotateTranscripts(TxDb.Hsapiens.UCSC.hg19.knownGene, "org.Hs.eg.db")
ICR_matchedGene <- matchGenes(ICRsGR, genes) #, type = "fiveprime"
mcols(ICRsGR) <- cbind(mcols(ICRsGR), ICR_matchedGene[, c(1, 3:12,14)])

# saveRDS(ICRsGR, file ="/home/qianhui/DNAme/Process_decidual/ICRs/ICRsGR.rds")
```


# 2. Plot beta values and ICRs (1st trimester samples)

The placenta-specific ICR were ICR46-59; ICR 1-45 were ubiquitously imprinted ICRs.

## 2.1 GRange for all probes (T1)
```{r}

colorFun <- colorRampPalette(brewer.pal(8,"Dark2"))
registerDoParallel(cores = 5)

table(colnames(NormBMlist$NormBAll) %in% GEO_phenotypes$Sample_name)
Matched_pd <- GEO_phenotypes[match(colnames(NormBMlist$NormBAll), GEO_phenotypes$Sample_name),]

PhenodataPlacenta <- Matched_pd[Matched_pd$Sample%in%"Placenta" & Matched_pd$Trimester%in%"First",]
# PhenodataPlacenta <- Matched_pd[Matched_pd$Trimester%in%"First",]

PhenodataPlacenta$Outlier <- factor(ifelse(PhenodataPlacenta$Additional_Info%in%"Outlier"| PhenodataPlacenta$Sample_name%in%c("GSM1702177_7970368142_R05C02", "GSM1947213_6008581028_R01C01"), "Outlier", "Standard"))


Placenta_Mnorm_v1 <- NormBMlist$NormMAll[,colnames(NormBMlist$NormMAll)%in%PhenodataPlacenta$Sample_name]
Placenta_betaNorm <- NormBMlist$NormBAll[,colnames(NormBMlist$NormBAll)%in%PhenodataPlacenta$Sample_name]

# some preperation for plotting
## Get probe ranges and select probe not related with islands
Gset_placenta <- makeGenomicRatioSetFromMatrix(Placenta_betaNorm, array = "IlluminaHumanMethylation450k",
                                               annotation = "ilmn12.hg19",mergeManifest = TRUE,
                                               what = "Beta")

GRangeProbe <- Gset_placenta@rowRanges[, c(1:2,7)]

# add mcol to GRangeProbe
## get GRange of all probes: 200k
### add values to GRange object metadata columns
matchWithGRprobe <- base::match(GRangeProbe$Name, rownames(Placenta_betaNorm))
values(GRangeProbe) <- cbind(values(GRangeProbe), DataFrame(Placenta_betaNorm[matchWithGRprobe,], check.names = F)) 

### add seqinfo
bs_hg19 <- BSgenome.Hsapiens.UCSC.hg19
bs_hg19
### check names are the same or not
seqnames(seqinfo(bs_hg19))[1:24] == seqinfo(GRangeProbe) %>% seqnames()

### assign seqinfo to my GRange for all probes 626374
seqinfo(GRangeProbe) <- Seqinfo(seqnames = seqnames(seqinfo(bs_hg19))[1:24],
                                seqlengths = seqlengths(seqinfo(bs_hg19))[1:24],
                                isCircular = isCircular(seqinfo(bs_hg19))[1:24],
                                genome = genome(seqinfo(bs_hg19))[1:24])

## get annotation and remove probes related wtih CpG islands
GRangeProbesForICRs <- GRangeProbe

```


## 2.2 GRange for probes (T2)
```{r}
# for Term samples

PhenodataPlacenta_T2 <- Matched_pd[Matched_pd$Sample%in%"Placenta" & Matched_pd$Trimester%in%"Second",]

PhenodataPlacenta_T2$Outlier <- factor(ifelse(PhenodataPlacenta_T2$Additional_Info%in%"Outlier"| PhenodataPlacenta_T2$Sample_name%in%c("GSM1702177_7970368142_R05C02", "GSM1947213_6008581028_R01C01"), "Outlier", "Standard"))


Placenta_Mnorm_v1_T2 <- NormBMlist$NormMAll[,colnames(NormBMlist$NormMAll)%in%PhenodataPlacenta_T2$Sample_name]
Placenta_betaNorm_T2 <- NormBMlist$NormBAll[,colnames(NormBMlist$NormBAll)%in%PhenodataPlacenta_T2$Sample_name]

# some preperation for plotting
## Get probe ranges and select probe not related with islands
Gset_placenta_T2 <- makeGenomicRatioSetFromMatrix(Placenta_betaNorm_T2, 
                                                    array = "IlluminaHumanMethylation450k",
                                               annotation = "ilmn12.hg19",mergeManifest = TRUE,
                                               what = "Beta")

GRangeProbe_T2 <- Gset_placenta_T2@rowRanges[, c(1:2,7)]

# add mcol to GRangeProbe
matchWithGRprobe_T2 <- base::match(GRangeProbe_T2$Name, rownames(Placenta_betaNorm_T2))
values(GRangeProbe_T2) <- cbind(values(GRangeProbe_T2), DataFrame(Placenta_betaNorm_T2[matchWithGRprobe_T2,], 
                                                            check.names = F)) 

### add seqinfo
### check names are the same or not
seqnames(seqinfo(bs_hg19))[1:24] == seqinfo(GRangeProbe_T2) %>% seqnames()

### assign seqinfo to my GRange for all probes 626374
seqinfo(GRangeProbe_T2) <- Seqinfo(seqnames = seqnames(seqinfo(bs_hg19))[1:24],
                                seqlengths = seqlengths(seqinfo(bs_hg19))[1:24],
                                isCircular = isCircular(seqinfo(bs_hg19))[1:24],
                                genome = genome(seqinfo(bs_hg19))[1:24])

## get annotation and remove probes related wtih CpG islands
GRangeProbesForICRs_T2 <- GRangeProbe_T2

```


## 2.3 GRange for probes (Term)
```{r}
# for Term samples

PhenodataPlacenta_Term <- Matched_pd[Matched_pd$Sample%in%"Placenta" & Matched_pd$Trimester%in%"Term",]

PhenodataPlacenta_Term$Outlier <- factor(ifelse(PhenodataPlacenta_Term$Additional_Info%in%"Outlier"| PhenodataPlacenta_Term$Sample_name%in%c("GSM1702177_7970368142_R05C02", "GSM1947213_6008581028_R01C01"), "Outlier", "Standard"))


Placenta_Mnorm_v1_Term <- NormBMlist$NormMAll[,colnames(NormBMlist$NormMAll)%in%PhenodataPlacenta_Term$Sample_name]
Placenta_betaNorm_Term <- NormBMlist$NormBAll[,colnames(NormBMlist$NormBAll)%in%PhenodataPlacenta_Term$Sample_name]

# some preperation for plotting
## Get probe ranges and select probe not related with islands
Gset_placenta_Term <- makeGenomicRatioSetFromMatrix(Placenta_betaNorm_Term, 
                                                    array = "IlluminaHumanMethylation450k",
                                               annotation = "ilmn12.hg19",mergeManifest = TRUE,
                                               what = "Beta")

GRangeProbe_Term <- Gset_placenta_Term@rowRanges[, c(1:2,7)]

# add mcol to GRangeProbe
matchWithGRprobe_Term <- base::match(GRangeProbe_Term$Name, rownames(Placenta_betaNorm_Term))
values(GRangeProbe_Term) <- cbind(values(GRangeProbe_Term), DataFrame(Placenta_betaNorm_Term[matchWithGRprobe_Term,], 
                                                            check.names = F)) 

### add seqinfo
### check names are the same or not
seqnames(seqinfo(bs_hg19))[1:24] == seqinfo(GRangeProbe_Term) %>% seqnames()

### assign seqinfo to my GRange for all probes 626374
seqinfo(GRangeProbe_Term) <- Seqinfo(seqnames = seqnames(seqinfo(bs_hg19))[1:24],
                                seqlengths = seqlengths(seqinfo(bs_hg19))[1:24],
                                isCircular = isCircular(seqinfo(bs_hg19))[1:24],
                                genome = genome(seqinfo(bs_hg19))[1:24])

## get annotation and remove probes related wtih CpG islands
GRangeProbesForICRs_Term <- GRangeProbe_Term

PhenodataPlacenta <- left_join(PhenodataPlacenta, placentaOutlier, by="Sample_name")
PhenodataPlacenta_T2 <- left_join(PhenodataPlacenta_T2, placentaOutlier, by="Sample_name")
PhenodataPlacenta_Term <- left_join(PhenodataPlacenta_Term, placentaOutlier, by="Sample_name")

```

## 2.3 Gviz plot for ICR (Chr9, GLIS3)

Note: The placenta-specific ICR were ICR46-59; ICR 1-45 were ubiquitously imprinted ICRs.

```{r}

# Gviz plot for each ICR
# 1:length(ICRsGR) #46
for(i in 46){

genome <- "hg19"

chrom <- as.character(seqnames(ICRsGR))[i]

start <- start(ICRsGR)[i]

end <- end(ICRsGR)[i]

# add extra space to plot (since plot range should be longer than this DMR)
minbase <- start - (0.01*(end-start))
maxbase <- end + (0.02*(end-start))

#### ideogram Track:
iTrack <- IdeogramTrack(genome = genome, chromosome = chrom, #name = "", 
                        from=minbase, to=maxbase,
                        fontsize=9, fontcolor="black")

#### Genome Axis Track:
gTrack <- GenomeAxisTrack(col="black", cex=0.7, name = "", fontcolor="black", add53=T, lwd=0.4,
                          # add35=T,
                          fontsize=7, col="black")

#### ICR track
ICRTrack <- AnnotationTrack(range=ICRsGR, gensome = genome, chromosome = chrom, 
                            start = start, end = end,
                            # name = paste0("ICR\n(GLIS3)",i),
                            id=paste0("ICR of ", ICRsGR$Gene[i]),
                            fontcolor.feature="black",
                            col.title="black",
                            #groupAnnotation=NULL,
                            background.title="transparent",
                            # showId=TRUE,
                            showFeatureId=TRUE,
                            stacking = "dense")

#### create methylation data Track:
library(RColorBrewer)

PhenodataPlacenta$Outlier2 <- str_replace_all(PhenodataPlacenta$Outlier, c("Outlier"="Impure", "Standard"="Pure"))
PhenodataPlacenta_T2$Outlier2 <- str_replace_all(PhenodataPlacenta_T2$Outlier, c("Outlier"="Impure", "Standard"="Pure"))
PhenodataPlacenta_Term$Outlier2 <- str_replace_all(PhenodataPlacenta_Term$Outlier, c("Outlier"="Impure", "Standard"="Pure"))

methTrack_T1 <- DataTrack(
  range = GRangeProbesForICRs[,4:24],
  genome = genome, chromosome = chrom, start = minbase, end = maxbase,
  # groups=factor(phenoDataPlacenta_GAordered$Gestational.Age),
  groups=factor(PhenodataPlacenta$Outlier3),
  ylim= c(0, 1.00),
  # col=unique(colorFun(18)[factor(phenoDataPlacenta_GAordered$Gestational.Age)]),
  # col=brewer.pal(8,"Dark2")[c(3,4,2,1)],
  # col=brewer.pal(8,"Dark2")[1:2],
  col=c(rep(brewer.pal(8,"Dark2")[1],5), brewer.pal(8,"Dark2")[2]),
  # type = c("a", "p"), # a: line plot of average
  # jitter.x=TRUE,
  type = c("b"), # a: line plot of average
  aggregateGroups=TRUE,
  name = "Beta value\n(First)",
  # background.panel = "black",
  fontsize.legend=9,
  # box.legend=TRUE,
  legend=FALSE,
  fontcolor.legend="black",
  background.title="transparent",
  col.title="black",
  col.axis="black",
  # cex.title=0.8,
  # cex.axis=0.8,
  cex.legend = 0.9
  )


methTrack_T2 <- DataTrack(
  range = GRangeProbesForICRs_T2[,4:19],
  genome = genome, chromosome = chrom, start = minbase, end = maxbase,
  # groups=factor(phenoDataPlacenta_GAordered$Gestational.Age),
  groups=factor(PhenodataPlacenta_T2$Outlier3),
  ylim= c(0, 1.00),
  # col=unique(colorFun(18)[factor(phenoDataPlacenta_GAordered$Gestational.Age)]),
  # col=brewer.pal(8,"Dark2")[c(3,4,2,1)],
  col=brewer.pal(8,"Dark2")[1:2],
  # type = c("a", "p"), # a: line plot of average
  # jitter.x=TRUE,
  type = c("b"), # a: line plot of average
  aggregateGroups=TRUE,
  name = "Beta value\n(Second)",
  # background.panel = "black",
  fontsize.legend=9,
  # box.legend=TRUE,
  legend=FALSE,
  fontcolor.legend="black",
  background.title="transparent",
  col.title="black",
  col.axis="black",
  # cex.title=0.8,
  # cex.axis=0.8,
  cex.legend = 0.9
  )


methTrack_Term <- DataTrack(
  range = GRangeProbesForICRs_Term[,4:345],
  genome = genome, chromosome = chrom, start = minbase, end = maxbase,
  # groups=factor(phenoDataPlacenta_GAordered$Gestational.Age),
  groups=factor(PhenodataPlacenta_Term$Outlier3),
  ylim= c(0, 1.00),
  # col=unique(colorFun(18)[factor(phenoDataPlacenta_GAordered$Gestational.Age)]),
  # col=brewer.pal(8,"Dark2")[c(3,4,2,1)],
  # col=brewer.pal(8,"Dark2")[1:2],
  col=c(rep(brewer.pal(8,"Dark2")[1],length(levels(factor(PhenodataPlacenta_Term$Outlier3)))-1),
        brewer.pal(8,"Dark2")[2]),
  type = c("b"), # a: line plot of average
  aggregateGroups=TRUE,
  # type = "smooth",  baseline=0.5, span=1/50,
  name = "Beta value\n(Term)",
  # background.panel = "black",
  fontsize.legend=9,
  legend=FALSE,
  # box.legend=TRUE,
  fontcolor.legend="black",
  background.title="transparent",
  col.title="black",
  col.axis="black",
  # cex.title=0.8,
  # cex.axis=0.8,
  cex.legend = 0.9
  )

# tracks to draw
tracks <- list(iTrack, gTrack, ICRTrack, methTrack_T1, methTrack_T2, methTrack_Term)

#### set up relative size for each track:
sizes <- c(1,1,1,3,3,3)
plotTracks(tracks, 
           from = minbase-1 , to = maxbase+1, 
           showTitle=TRUE,
           add53 = TRUE, 
           # add35 = TRUE, 
           grid=FALSE,
           # lty.grid=3,
           sizes = sizes,
           #showBandId=TRUE,
           #showId=TRUE,
           #showFeatureId=TRUE,
           length(tracks))

tiff(filename = paste0("/home/qianhui/DNAme/Process_decidual/ICRs/ICRfigures/Gviz_T1Termsamples_sub_ICR_", i, ".tiff"), 
     width = 12, height = 8, units = "cm", res = 300)
plotTracks(tracks, 
           from = minbase-1 , to = maxbase+1, 
           showTitle=TRUE,
           add53 = TRUE, 
           # add35 = TRUE, 
           grid=FALSE,
           # lty.grid=3,
           sizes = sizes,
           #showBandId=TRUE,
           #showId=TRUE,
           #showFeatureId=TRUE,
           length(tracks))
dev.off()
}

```

## 2.4 plot violin plot for placenta specific ICRs


```{r}
# T1 samples
ProbesBetaInICR <- find_overlaps(ICRsGR[ICRsGR$ICR_Type%in%"PlacentaSpecific",], GRangeProbesForICRs) %>% 
  plyranges::select(Gene, Name, GSM1616998_8795207029_R02C01:`201414140063_R08C01`)%>% 
  mcols() %>% unique() %>% as.data.frame() %>% 
  remove_rownames() #%>% column_to_rownames(var="Name")

# T2 samples
ProbesBetaInICR_T2 <- find_overlaps(ICRsGR[ICRsGR$ICR_Type%in%"PlacentaSpecific",], GRangeProbesForICRs_T2) %>% 
  plyranges::select(Gene, Name, GSM1702082_7970368015_R03C01:GSM1702246_9406922117_R04C01)%>% 
  mcols() %>% unique() %>% as.data.frame() %>% 
  remove_rownames() #%>% column_to_rownames(var="Name")

# Term samples
ProbesBetaInICR_Term <- find_overlaps(ICRsGR[ICRsGR$ICR_Type%in%"PlacentaSpecific",], GRangeProbesForICRs_Term) %>% 
  plyranges::select(Gene, Name, GSM1931568_9285451059_R06C02:GSM3179755_200925700033_R04C01)%>% 
  mcols() %>% unique() %>% as.data.frame() %>% 
  remove_rownames() #%>% column_to_rownames(var="Name")

# join T1 and Term samples
ProbesBetaInICR_T1T2Term <- ProbesBetaInICR %>% 
  left_join(ProbesBetaInICR_T2, by="Name") %>% 
  left_join(ProbesBetaInICR_Term, by="Name") %>% 
  dplyr::select(-Gene.x, -Gene.y)

## change matrix to a long data frame
ProbesBetaInICR_long <- ProbesBetaInICR_T1T2Term %>% as.data.frame() %>% 
  gather(key = "sampleName",
         value = "Betas",
         # contains("mean_")
         -c(Name, Gene)
         )

## add outlier information as a column
ProbesBetaInICR_long_v1 <- ProbesBetaInICR_long %>% mutate(sampleName= str_replace_all(sampleName, "^X", ""))
# ProbesBetaInICR_long_v1 <- ProbesBetaInICR_long %>% mutate(sampleName= str_replace_all(sampleName, "mean_", ""))

PhenodataPlacentaT1T2Term <- rbind(PhenodataPlacenta, PhenodataPlacenta_T2, PhenodataPlacenta_Term)

PhenodataPlacentaT1T2Term_v1 <- PhenodataPlacentaT1T2Term %>% mutate(sampleName=Sample_name) %>% 
  dplyr::select(sampleName, Outlier, Trimester)

ProbesBetaInICR_long_v2 <- left_join(ProbesBetaInICR_long_v1, PhenodataPlacentaT1T2Term_v1, by="sampleName")

#color setting
pal <- brewer.pal(8, "Dark2")
pal2 <- c("#08519C",pal[2])
namedPal2 <- pal2 %>% `names<-`(unique(ProbesBetaInICR_long_v2$Outliers))

# plot and save the violin plot
ICRsViolinPlot <- ggplot(ProbesBetaInICR_long_v2, aes(Outlier, Betas, fill=Outlier))+ 
  geom_violin(trim = T, aes(fill=Outlier))+ 
  scale_fill_brewer(palette = "Dark2", labels=c("Outlier"="Mixed", "Standard"="Pure"))+
  labs(x="", y="Beta values", fill="Placenta")+
  geom_boxplot(width=0.1, fill="white")+
  facet_wrap(~Trimester)+
  stat_compare_means(label = "p.signif", method = "wilcox.test") +
  theme(text = element_text(size=10, colour = "black"),
        # axis.text.x=element_text(size=6, colour = "black"),
        axis.text.x=element_blank(),
        axis.text.y=element_text(size=10, colour = "black"),
        legend.position = "right", 
        strip.text.x = element_text(size = 10, colour = "black"))

tiff(file = "/home/qianhui/DNAme/Process_decidual/ICRs/ICRfigures/ICRprobes_violinPlot_374Placenta_sub.tiff", 
     width = 10, height = 7, units = "cm", res = 300)
theme_set(theme_pubr())
ICRsViolinPlot
dev.off()

saveRDS(ICRsViolinPlot, file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/allICRsViolinPlot_sub.rds")

```

Statistical test using Mannâ€“Whitney U test:

```{r}
# statistic test

for(i in c("First", "Second", "Term")){
wilcox.test(ProbesBetaInICR_long_v2[ProbesBetaInICR_long_v2$Trimester%in%i, ]$Betas
            ~ProbesBetaInICR_long_v2[ProbesBetaInICR_long_v2$Trimester%in%i, ]$Outlier) %>% 
    print()
}

```


### plot T1, T2 & term seperatly, facet_wrap by Trimester (Sup. Figure 2)
```{r}

T1T2TermICRGenes<- ggplot(ProbesBetaInICR_long_v2,  #[ProbesBetaInICR_long_v2$Trimester%in%"First",]
       aes(Gene, Betas, fill=Outlier))+ 
  # geom_violin(trim = T, aes(fill=Outlier))+ 
  geom_boxplot(aes(fill=Outlier))+ 
  scale_fill_brewer(palette = "Dark2", labels=c("Outlier"="Mixed", "Standard"="Pure"))+
  labs(x="", y="Beta values", fill="Placenta")+
  facet_wrap(~Trimester, nrow = 3, ncol = 1)+
  stat_compare_means(label = "p.signif", method = "wilcox.test") +
  theme(text = element_text(size=10, colour = "black"),
        axis.text.x=element_text(size=9, angle = 45, vjust = 0.9, hjust = 0.95, colour = "black"),
        # axis.text.x=element_blank(),
        axis.text.y=element_text(size=10, colour = "black"),
        legend.position = "right", 
        strip.text.x = element_text(size = 10, colour = "black"))

tiff(file = "/home/qianhui/DNAme/Process_decidual/ICRs/ICRfigures/ICRgenes_violinPlot_T1TermPlacenta_v1_sub.tiff", 
     width = 25, height = 17, units = "cm", res = 300)
theme_set(theme_pubr())
T1T2TermICRGenes
dev.off()

T1T2TermICRGenes

```



# 3. arrange violin plot of PMD and ICR together (Figure2 A/B)

```{r}

AllPMDsMeanViolinPlot <- readRDS(file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/AllPMDsMeanViolinPlot_sub.rds")
ICRsViolinPlot <- readRDS(file = "/home/qianhui/DNAme/Process_decidual/RDSfiles/allICRsViolinPlot_sub.rds")

AllPMDsMeanViolinPlot <- AllPMDsMeanViolinPlot+ggtitle("DNA methylation in PMDs")+theme(plot.title = element_text(hjust = 0.5))

ICRsViolinPlot <- ICRsViolinPlot+ggtitle("DNA methylation in ICRs")+theme(plot.title = element_text(hjust = 0.5))
  
tiff(file = "/home/qianhui/DNAme/Process_decidual/figures/TechFigure2AB_sub.tiff",
     width = 25, height = 7, units = "cm", res = 300)
theme_set(theme_pubr())

ggarrange(AllPMDsMeanViolinPlot, ICRsViolinPlot, 
          labels = c("A", "B"),
          font.label = list(size = 12, face = "bold", color ="black"),
          ncol = 2)
dev.off()

```


